{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "N0zivH3dLry7"
   },
   "source": [
    "### **Prerequesites**: Download uniprot data and import python libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "id": "XoafbphKAjkC"
   },
   "outputs": [],
   "source": [
    "!wget https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz\n",
    "!gunzip uniprot_sprot.dat.gz\n",
    "import re\n",
    "from collections import defaultdict\n",
    "import torch\n",
    "import pickle\n",
    "import torch.nn as nn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID   001R_FRG3G              Reviewed;         256 AA.\n",
      "AC   Q6GZX4;\n",
      "DT   28-JUN-2011, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   19-JUL-2004, sequence version 1.\n",
      "DT   09-APR-2025, entry version 45.\n",
      "DE   RecName: Full=Putative transcription factor 001R;\n",
      "GN   ORFNames=FV3-001R;\n",
      "OS   Frog virus 3 (isolate Goorha) (FV-3).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Alphairidovirinae; Ranavirus; Frog virus 3.\n",
      "OX   NCBI_TaxID=654924;\n",
      "OH   NCBI_TaxID=30343; Dryophytes versicolor (chameleon treefrog).\n",
      "OH   NCBI_TaxID=8404; Lithobates pipiens (Northern leopard frog) (Rana pipiens).\n",
      "OH   NCBI_TaxID=45438; Lithobates sylvaticus (Wood frog) (Rana sylvatica).\n",
      "OH   NCBI_TaxID=8316; Notophthalmus viridescens (Eastern newt) (Triturus viridescens).\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=15165820; DOI=10.1016/j.virol.2004.02.019;\n",
      "RA   Tan W.G., Barkman T.J., Gregory Chinchar V., Essani K.;\n",
      "RT   \"Comparative genomic analyses of frog virus 3, type species of the genus\n",
      "RT   Ranavirus (family Iridoviridae).\";\n",
      "RL   Virology 323:70-84(2004).\n",
      "CC   -!- FUNCTION: Transcription activation. {ECO:0000305}.\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; AY548484; AAT09660.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_031579.1; NC_005946.1.\n",
      "DR   SwissPalm; Q6GZX4; -.\n",
      "DR   KEGG; vg:2947773; -.\n",
      "DR   Proteomes; UP000008770; Segment.\n",
      "DR   GO; GO:0046782; P:regulation of viral transcription; IEA:InterPro.\n",
      "DR   InterPro; IPR007031; Poxvirus_VLTF3.\n",
      "DR   Pfam; PF04947; Pox_VLTF3; 1.\n",
      "PE   4: Predicted;\n",
      "KW   Activator; Reference proteome; Transcription; Transcription regulation.\n",
      "FT   CHAIN           1..256\n",
      "FT                   /note=\"Putative transcription factor 001R\"\n",
      "FT                   /id=\"PRO_0000410512\"\n",
      "SQ   SEQUENCE   256 AA;  29735 MW;  B4840739BF7D4121 CRC64;\n",
      "MAFSAEDVLK EYDRRRRMEA LLLSLYYPND RKLLDYKEWS PPRVQVECPK APVEWNNPPS\n",
      "EKGLIVGHFS GIKYKGEKAQ ASEVDVNKMC CWVSKFKDAM RRYQGIQTCK IPGKVLSDLD\n",
      "AKIKAYNLTV EGVEGFVRYS RVTKQHVAAF LKELRHSKQY ENVNLIHYIL TDKRVDIQHL\n",
      "EKDLVKDFKA LVESAHRMRQ GHMINVKYIL YQLLKKHGHG PDGPDILTVK TGSKGVLYDD\n",
      "SFRKIYTDLG WKFTPL\n",
      "//\n",
      "ID   002L_FRG3G              Reviewed;         320 AA.\n",
      "AC   Q6GZX3;\n",
      "DT   28-JUN-2011, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   19-JUL-2004, sequence version 1.\n",
      "DT   09-APR-2025, entry version 48.\n",
      "DE   RecName: Full=Uncharacterized protein 002L;\n",
      "GN   ORFNames=FV3-002L;\n",
      "OS   Frog virus 3 (isolate Goorha) (FV-3).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Alphairidovirinae; Ranavirus; Frog virus 3.\n",
      "OX   NCBI_TaxID=654924;\n",
      "OH   NCBI_TaxID=30343; Dryophytes versicolor (chameleon treefrog).\n",
      "OH   NCBI_TaxID=8404; Lithobates pipiens (Northern leopard frog) (Rana pipiens).\n",
      "OH   NCBI_TaxID=45438; Lithobates sylvaticus (Wood frog) (Rana sylvatica).\n",
      "OH   NCBI_TaxID=8316; Notophthalmus viridescens (Eastern newt) (Triturus viridescens).\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=15165820; DOI=10.1016/j.virol.2004.02.019;\n",
      "RA   Tan W.G., Barkman T.J., Gregory Chinchar V., Essani K.;\n",
      "RT   \"Comparative genomic analyses of frog virus 3, type species of the genus\n",
      "RT   Ranavirus (family Iridoviridae).\";\n",
      "RL   Virology 323:70-84(2004).\n",
      "CC   -!- SUBCELLULAR LOCATION: Host membrane {ECO:0000305}; Single-pass membrane\n",
      "CC       protein {ECO:0000305}.\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; AY548484; AAT09661.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_031580.1; NC_005946.1.\n",
      "DR   KEGG; vg:2947774; -.\n",
      "DR   Proteomes; UP000008770; Segment.\n",
      "DR   GO; GO:0033644; C:host cell membrane; IEA:UniProtKB-SubCell.\n",
      "DR   GO; GO:0016020; C:membrane; IEA:UniProtKB-KW.\n",
      "DR   InterPro; IPR004251; Pox_virus_G9/A16.\n",
      "DR   Pfam; PF03003; Pox_G9-A16; 1.\n",
      "PE   4: Predicted;\n",
      "KW   Host membrane; Membrane; Reference proteome; Transmembrane;\n",
      "KW   Transmembrane helix.\n",
      "FT   CHAIN           1..320\n",
      "FT                   /note=\"Uncharacterized protein 002L\"\n",
      "FT                   /id=\"PRO_0000410509\"\n",
      "FT   TRANSMEM        301..318\n",
      "FT                   /note=\"Helical\"\n",
      "FT                   /evidence=\"ECO:0000255\"\n",
      "FT   REGION          261..294\n",
      "FT                   /note=\"Disordered\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "FT   COMPBIAS        262..294\n",
      "FT                   /note=\"Pro residues\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "SQ   SEQUENCE   320 AA;  34642 MW;  9E110808B6E328E0 CRC64;\n",
      "MSIIGATRLQ NDKSDTYSAG PCYAGGCSAF TPRGTCGKDW DLGEQTCASG FCTSQPLCAR\n",
      "IKKTQVCGLR YSSKGKDPLV SAEWDSRGAP YVRCTYDADL IDTQAQVDQF VSMFGESPSL\n",
      "AERYCMRGVK NTAGELVSRV SSDADPAGGW CRKWYSAHRG PDQDAALGSF CIKNPGAADC\n",
      "KCINRASDPV YQKVKTLHAY PDQCWYVPCA ADVGELKMGT QRDTPTNCPT QVCQIVFNML\n",
      "DDGSVTMDDV KNTINCDFSK YVPPPPPPKP TPPTPPTPPT PPTPPTPPTP PTPRPVHNRK\n",
      "VMFFVAGAVL VAILISTVRW\n",
      "//\n",
      "ID   002R_IIV3               Reviewed;         458 AA.\n",
      "AC   Q197F8;\n",
      "DT   16-JUN-2009, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   11-JUL-2006, sequence version 1.\n",
      "DT   09-APR-2025, entry version 31.\n",
      "DE   RecName: Full=Uncharacterized protein 002R;\n",
      "GN   ORFNames=IIV3-002R;\n",
      "OS   Invertebrate iridescent virus 3 (IIV-3) (Mosquito iridescent virus).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Betairidovirinae; Chloriridovirus.\n",
      "OX   NCBI_TaxID=345201;\n",
      "OH   NCBI_TaxID=7163; Aedes vexans (Inland floodwater mosquito) (Culex vexans).\n",
      "OH   NCBI_TaxID=42431; Culex territans.\n",
      "OH   NCBI_TaxID=332058; Culiseta annulata.\n",
      "OH   NCBI_TaxID=310513; Ochlerotatus sollicitans (eastern saltmarsh mosquito).\n",
      "OH   NCBI_TaxID=329105; Ochlerotatus taeniorhynchus (Black salt marsh mosquito) (Aedes taeniorhynchus).\n",
      "OH   NCBI_TaxID=7183; Psorophora ferox.\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=16912294; DOI=10.1128/jvi.00464-06;\n",
      "RA   Delhon G., Tulman E.R., Afonso C.L., Lu Z., Becnel J.J., Moser B.A.,\n",
      "RA   Kutish G.F., Rock D.L.;\n",
      "RT   \"Genome of invertebrate iridescent virus type 3 (mosquito iridescent\n",
      "RT   virus).\";\n",
      "RL   J. Virol. 80:8439-8449(2006).\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; DQ643392; ABF82032.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_654574.1; NC_008187.1.\n",
      "DR   KEGG; vg:4156251; -.\n",
      "DR   Proteomes; UP000001358; Genome.\n",
      "PE   4: Predicted;\n",
      "KW   Reference proteome.\n",
      "FT   CHAIN           1..458\n",
      "FT                   /note=\"Uncharacterized protein 002R\"\n",
      "FT                   /id=\"PRO_0000377938\"\n",
      "FT   REGION          339..397\n",
      "FT                   /note=\"Disordered\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "FT   REGION          434..458\n",
      "FT                   /note=\"Disordered\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "FT   COMPBIAS        344..390\n",
      "FT                   /note=\"Acidic residues\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "FT   COMPBIAS        436..458\n",
      "FT                   /note=\"Acidic residues\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "SQ   SEQUENCE   458 AA;  53921 MW;  E46E5C85D7ACA139 CRC64;\n",
      "MASNTVSAQG GSNRPVRDFS NIQDVAQFLL FDPIWNEQPG SIVPWKMNRE QALAERYPEL\n",
      "QTSEPSEDYS GPVESLELLP LEIKLDIMQY LSWEQISWCK HPWLWTRWYK DNVVRVSAIT\n",
      "FEDFQREYAF PEKIQEIHFT DTRAEEIKAI LETTPNVTRL VIRRIDDMNY NTHGDLGLDD\n",
      "LEFLTHLMVE DACGFTDFWA PSLTHLTIKN LDMHPRWFGP VMDGIKSMQS TLKYLYIFET\n",
      "YGVNKPFVQW CTDNIETFYC TNSYRYENVP RPIYVWVLFQ EDEWHGYRVE DNKFHRRYMY\n",
      "STILHKRDTD WVENNPLKTP AQVEMYKFLL RISQLNRDGT GYESDSDPEN EHFDDESFSS\n",
      "GEEDSSDEDD PTWAPDSDDS DWETETEEEP SVAARILEKG KLTITNLMKS LGFKPKPKKI\n",
      "QSIDRYFCSL DSNYNSEDED FEYDSDSEDD DSDSEDDC\n",
      "//\n",
      "ID   003L_IIV3               Reviewed;         156 AA.\n",
      "AC   Q197F7;\n",
      "DT   16-JUN-2009, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   11-JUL-2006, sequence version 1.\n",
      "DT   09-APR-2025, entry version 24.\n",
      "DE   RecName: Full=Uncharacterized protein 003L;\n",
      "GN   ORFNames=IIV3-003L;\n",
      "OS   Invertebrate iridescent virus 3 (IIV-3) (Mosquito iridescent virus).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Betairidovirinae; Chloriridovirus.\n",
      "OX   NCBI_TaxID=345201;\n",
      "OH   NCBI_TaxID=7163; Aedes vexans (Inland floodwater mosquito) (Culex vexans).\n",
      "OH   NCBI_TaxID=42431; Culex territans.\n",
      "OH   NCBI_TaxID=332058; Culiseta annulata.\n",
      "OH   NCBI_TaxID=310513; Ochlerotatus sollicitans (eastern saltmarsh mosquito).\n",
      "OH   NCBI_TaxID=329105; Ochlerotatus taeniorhynchus (Black salt marsh mosquito) (Aedes taeniorhynchus).\n",
      "OH   NCBI_TaxID=7183; Psorophora ferox.\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=16912294; DOI=10.1128/jvi.00464-06;\n",
      "RA   Delhon G., Tulman E.R., Afonso C.L., Lu Z., Becnel J.J., Moser B.A.,\n",
      "RA   Kutish G.F., Rock D.L.;\n",
      "RT   \"Genome of invertebrate iridescent virus type 3 (mosquito iridescent\n",
      "RT   virus).\";\n",
      "RL   J. Virol. 80:8439-8449(2006).\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; DQ643392; ABF82033.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_654575.1; NC_008187.1.\n",
      "DR   KEGG; vg:4156252; -.\n",
      "DR   Proteomes; UP000001358; Genome.\n",
      "PE   4: Predicted;\n",
      "KW   Reference proteome.\n",
      "FT   CHAIN           1..156\n",
      "FT                   /note=\"Uncharacterized protein 003L\"\n",
      "FT                   /id=\"PRO_0000377939\"\n",
      "SQ   SEQUENCE   156 AA;  17043 MW;  D48A43940FF8C815 CRC64;\n",
      "MYQAINPCPQ SWYGSPQLER EIVCKMSGAP HYPNYYPVHP NALGGAWFDT SLNARSLTTT\n",
      "PSLTTCTPPS LAACTPPTSL GMVDSPPHIN PPRRIGTLCF DFGSAKSPQR CECVASDRPS\n",
      "TTSNTAPDTY RLLITNSKTR KNNYGTCRLE PLTYGI\n",
      "//\n",
      "ID   003R_FRG3G              Reviewed;         438 AA.\n",
      "AC   Q6GZX2;\n",
      "DT   28-JUN-2011, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   19-JUL-2004, sequence version 1.\n",
      "DT   09-APR-2025, entry version 39.\n",
      "DE   RecName: Full=Uncharacterized protein 3R;\n",
      "DE   Flags: Precursor;\n",
      "GN   ORFNames=FV3-003R;\n",
      "OS   Frog virus 3 (isolate Goorha) (FV-3).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Alphairidovirinae; Ranavirus; Frog virus 3.\n",
      "OX   NCBI_TaxID=654924;\n",
      "OH   NCBI_TaxID=30343; Dryophytes versicolor (chameleon treefrog).\n",
      "OH   NCBI_TaxID=8404; Lithobates pipiens (Northern leopard frog) (Rana pipiens).\n",
      "OH   NCBI_TaxID=45438; Lithobates sylvaticus (Wood frog) (Rana sylvatica).\n",
      "OH   NCBI_TaxID=8316; Notophthalmus viridescens (Eastern newt) (Triturus viridescens).\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=15165820; DOI=10.1016/j.virol.2004.02.019;\n",
      "RA   Tan W.G., Barkman T.J., Gregory Chinchar V., Essani K.;\n",
      "RT   \"Comparative genomic analyses of frog virus 3, type species of the genus\n",
      "RT   Ranavirus (family Iridoviridae).\";\n",
      "RL   Virology 323:70-84(2004).\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; AY548484; AAT09662.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_031581.1; NC_005946.1.\n",
      "DR   KEGG; vg:2947775; -.\n",
      "DR   Proteomes; UP000008770; Segment.\n",
      "PE   3: Inferred from homology;\n",
      "KW   Reference proteome; Signal.\n",
      "FT   SIGNAL          1..32\n",
      "FT                   /evidence=\"ECO:0000255\"\n",
      "FT   CHAIN           33..438\n",
      "FT                   /note=\"Uncharacterized protein 3R\"\n",
      "FT                   /id=\"PRO_0000410532\"\n",
      "SQ   SEQUENCE   438 AA;  48297 MW;  075C8FA17B3C5C56 CRC64;\n",
      "MARPLLGKTS SVRRRLESLS ACSIFFFLRK FCQKMASLVF LNSPVYQMSN ILLTERRQVD\n",
      "RAMGGSDDDG VMVVALSPSD FKTVLGSALL AVERDMVHVV PKYLQTPGIL HDMLVLLTPI\n",
      "FGEALSVDMS GATDVMVQQI ATAGFVDVDP LHSSVSWKDN VSCPVALLAV SNAVRTMMGQ\n",
      "PCQVTLIIDV GTQNILRDLV NLPVEMSGDL QVMAYTKDPL GKVPAVGVSV FDSGSVQKGD\n",
      "AHSVGAPDGL VSFHTHPVSS AVELNYHAGW PSNVDMSSLL TMKNLMHVVV AEEGLWTMAR\n",
      "TLSMQRLTKV LTDAEKDVMR AAAFNLFLPL NELRVMGTKD SNNKSLKTYF EVFETFTIGA\n",
      "LMKHSGVTPT AFVDRRWLDN TIYHMGFIPW GRDMRFVVEY DLDGTNPFLN TVPTLMSVKR\n",
      "KAKIQEMFDN MVSRMVTS\n",
      "//\n",
      "ID   004R_FRG3G              Reviewed;          60 AA.\n",
      "AC   Q6GZX1;\n",
      "DT   28-JUN-2011, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   19-JUL-2004, sequence version 1.\n",
      "DT   09-APR-2025, entry version 41.\n",
      "DE   RecName: Full=Uncharacterized protein 004R;\n",
      "GN   ORFNames=FV3-004R;\n",
      "OS   Frog virus 3 (isolate Goorha) (FV-3).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Alphairidovirinae; Ranavirus; Frog virus 3.\n",
      "OX   NCBI_TaxID=654924;\n",
      "OH   NCBI_TaxID=30343; Dryophytes versicolor (chameleon treefrog).\n",
      "OH   NCBI_TaxID=8404; Lithobates pipiens (Northern leopard frog) (Rana pipiens).\n",
      "OH   NCBI_TaxID=45438; Lithobates sylvaticus (Wood frog) (Rana sylvatica).\n",
      "OH   NCBI_TaxID=8316; Notophthalmus viridescens (Eastern newt) (Triturus viridescens).\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=15165820; DOI=10.1016/j.virol.2004.02.019;\n",
      "RA   Tan W.G., Barkman T.J., Gregory Chinchar V., Essani K.;\n",
      "RT   \"Comparative genomic analyses of frog virus 3, type species of the genus\n",
      "RT   Ranavirus (family Iridoviridae).\";\n",
      "RL   Virology 323:70-84(2004).\n",
      "CC   -!- SUBCELLULAR LOCATION: Host membrane {ECO:0000305}; Single-pass membrane\n",
      "CC       protein {ECO:0000305}.\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; AY548484; AAT09663.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_031582.1; NC_005946.1.\n",
      "DR   SMR; Q6GZX1; -.\n",
      "DR   KEGG; vg:2947776; -.\n",
      "DR   Proteomes; UP000008770; Segment.\n",
      "DR   GO; GO:0033644; C:host cell membrane; IEA:UniProtKB-SubCell.\n",
      "DR   GO; GO:0016020; C:membrane; IEA:UniProtKB-KW.\n",
      "PE   4: Predicted;\n",
      "KW   Host membrane; Membrane; Reference proteome; Transmembrane;\n",
      "KW   Transmembrane helix.\n",
      "FT   CHAIN           1..60\n",
      "FT                   /note=\"Uncharacterized protein 004R\"\n",
      "FT                   /id=\"PRO_0000410528\"\n",
      "FT   TRANSMEM        14..34\n",
      "FT                   /note=\"Helical\"\n",
      "FT                   /evidence=\"ECO:0000255\"\n",
      "FT   REGION          38..60\n",
      "FT                   /note=\"Disordered\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "FT   COMPBIAS        40..60\n",
      "FT                   /note=\"Polar residues\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "SQ   SEQUENCE   60 AA;  6514 MW;  12F072778EE6DFE4 CRC64;\n",
      "MNAKYDTDQG VGRMLFLGTI GLAVVVGGLM AYGYYYDGKT PSSGTSFHTA SPSFSSRYRY\n",
      "//\n",
      "ID   005L_IIV3               Reviewed;         217 AA.\n",
      "AC   Q197F5;\n",
      "DT   16-JUN-2009, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   11-JUL-2006, sequence version 1.\n",
      "DT   09-APR-2025, entry version 33.\n",
      "DE   RecName: Full=Uncharacterized protein 005L;\n",
      "DE   Flags: Precursor;\n",
      "GN   ORFNames=IIV3-005L;\n",
      "OS   Invertebrate iridescent virus 3 (IIV-3) (Mosquito iridescent virus).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Betairidovirinae; Chloriridovirus.\n",
      "OX   NCBI_TaxID=345201;\n",
      "OH   NCBI_TaxID=7163; Aedes vexans (Inland floodwater mosquito) (Culex vexans).\n",
      "OH   NCBI_TaxID=42431; Culex territans.\n",
      "OH   NCBI_TaxID=332058; Culiseta annulata.\n",
      "OH   NCBI_TaxID=310513; Ochlerotatus sollicitans (eastern saltmarsh mosquito).\n",
      "OH   NCBI_TaxID=329105; Ochlerotatus taeniorhynchus (Black salt marsh mosquito) (Aedes taeniorhynchus).\n",
      "OH   NCBI_TaxID=7183; Psorophora ferox.\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=16912294; DOI=10.1128/jvi.00464-06;\n",
      "RA   Delhon G., Tulman E.R., Afonso C.L., Lu Z., Becnel J.J., Moser B.A.,\n",
      "RA   Kutish G.F., Rock D.L.;\n",
      "RT   \"Genome of invertebrate iridescent virus type 3 (mosquito iridescent\n",
      "RT   virus).\";\n",
      "RL   J. Virol. 80:8439-8449(2006).\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; DQ643392; ABF82035.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_654577.1; NC_008187.1.\n",
      "DR   KEGG; vg:4156254; -.\n",
      "DR   Proteomes; UP000001358; Genome.\n",
      "PE   3: Inferred from homology;\n",
      "KW   Reference proteome; Signal.\n",
      "FT   SIGNAL          1..24\n",
      "FT                   /evidence=\"ECO:0000255\"\n",
      "FT   CHAIN           25..217\n",
      "FT                   /note=\"Uncharacterized protein 005L\"\n",
      "FT                   /id=\"PRO_0000377940\"\n",
      "SQ   SEQUENCE   217 AA;  23854 MW;  6DBE65967D7EEBD5 CRC64;\n",
      "MRYTVLIALQ GALLLLLLID DGQGQSPYPY PGMPCNSSRQ CGLGTCVHSR CAHCSSDGTL\n",
      "CSPEDPTMVW PCCPESSCQL VVGLPSLVNH YNCLPNQCTD SSQCPGGFGC MTRRSKCELC\n",
      "KADGEACNSP YLDWRKDKEC CSGYCHTEAR GLEGVCIDPK KIFCTPKNPW QLAPYPPSYH\n",
      "QPTTLRPPTS LYDSWLMSGF LVKSTTAPST QEEEDDY\n",
      "//\n",
      "ID   005R_FRG3G              Reviewed;         204 AA.\n",
      "AC   Q6GZX0;\n",
      "DT   28-JUN-2011, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   19-JUL-2004, sequence version 1.\n",
      "DT   09-APR-2025, entry version 49.\n",
      "DE   RecName: Full=Uncharacterized protein 005R;\n",
      "GN   ORFNames=FV3-005R;\n",
      "OS   Frog virus 3 (isolate Goorha) (FV-3).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Alphairidovirinae; Ranavirus; Frog virus 3.\n",
      "OX   NCBI_TaxID=654924;\n",
      "OH   NCBI_TaxID=30343; Dryophytes versicolor (chameleon treefrog).\n",
      "OH   NCBI_TaxID=8404; Lithobates pipiens (Northern leopard frog) (Rana pipiens).\n",
      "OH   NCBI_TaxID=45438; Lithobates sylvaticus (Wood frog) (Rana sylvatica).\n",
      "OH   NCBI_TaxID=8316; Notophthalmus viridescens (Eastern newt) (Triturus viridescens).\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=15165820; DOI=10.1016/j.virol.2004.02.019;\n",
      "RA   Tan W.G., Barkman T.J., Gregory Chinchar V., Essani K.;\n",
      "RT   \"Comparative genomic analyses of frog virus 3, type species of the genus\n",
      "RT   Ranavirus (family Iridoviridae).\";\n",
      "RL   Virology 323:70-84(2004).\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; AY548484; AAT09664.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_031583.1; NC_005946.1.\n",
      "DR   KEGG; vg:2947777; -.\n",
      "DR   Proteomes; UP000008770; Segment.\n",
      "DR   InterPro; IPR003360; US22-like.\n",
      "DR   Pfam; PF02393; US22; 1.\n",
      "PE   4: Predicted;\n",
      "KW   Reference proteome.\n",
      "FT   CHAIN           1..204\n",
      "FT                   /note=\"Uncharacterized protein 005R\"\n",
      "FT                   /id=\"PRO_0000410513\"\n",
      "FT   REGION          1..20\n",
      "FT                   /note=\"Disordered\"\n",
      "FT                   /evidence=\"ECO:0000256|SAM:MobiDB-lite\"\n",
      "SQ   SEQUENCE   204 AA;  23418 MW;  13AC9A179015F769 CRC64;\n",
      "MQNPLPEVMS PEHDKRTTTP MSKEANKFIR ELDKKPGDLA VVSDFVKRNT GKRLPIGKRS\n",
      "NLYVRICDLS GTIYMGETFI LESWEELYLP EPTKMEVLGT LESCCGIPPF PEWIVMVGED\n",
      "QCVYAYGDEE ILLFAYSVKQ LVEEGIQETG ISYKYPDDIS DVDEEVLQQD EEIQKIRKKT\n",
      "REFVDKDAQE FQDFLNSLDA SLLS\n",
      "//\n",
      "ID   006L_IIV6               Reviewed;         352 AA.\n",
      "AC   Q91G88;\n",
      "DT   16-JUN-2009, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   01-DEC-2001, sequence version 1.\n",
      "DT   09-APR-2025, entry version 54.\n",
      "DE   RecName: Full=Putative KilA-N domain-containing protein 006L;\n",
      "GN   ORFNames=IIV6-006L;\n",
      "OS   Invertebrate iridescent virus 6 (IIV-6) (Chilo iridescent virus).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Betairidovirinae; Iridovirus.\n",
      "OX   NCBI_TaxID=176652;\n",
      "OH   NCBI_TaxID=6997; Acheta domesticus (House cricket).\n",
      "OH   NCBI_TaxID=168631; Chilo suppressalis (Asiatic rice borer moth).\n",
      "OH   NCBI_TaxID=6999; Gryllus bimaculatus (Two-spotted cricket).\n",
      "OH   NCBI_TaxID=58607; Gryllus campestris.\n",
      "OH   NCBI_TaxID=7108; Spodoptera frugiperda (Fall armyworm).\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=11448171; DOI=10.1006/viro.2001.0963;\n",
      "RA   Jakob N.J., Mueller K., Bahr U., Darai G.;\n",
      "RT   \"Analysis of the first complete DNA sequence of an invertebrate iridovirus:\n",
      "RT   coding strategy of the genome of Chilo iridescent virus.\";\n",
      "RL   Virology 286:182-196(2001).\n",
      "RN   [2]\n",
      "RP   GENOME REANNOTATION.\n",
      "RX   PubMed=17239238; DOI=10.1186/1743-422x-4-11;\n",
      "RA   Eaton H.E., Metcalf J., Penny E., Tcherepanov V., Upton C., Brunetti C.R.;\n",
      "RT   \"Comparative genomic analysis of the family Iridoviridae: re-annotating and\n",
      "RT   defining the core set of iridovirus genes.\";\n",
      "RL   Virol. J. 4:11-11(2007).\n",
      "CC   -!- SIMILARITY: Belongs to the IIV-6 006L/238R/313L/468L family.\n",
      "CC       {ECO:0000305}.\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; AF303741; AAK81943.1; -; Genomic_DNA.\n",
      "DR   RefSeq; NP_149469.1; NC_003038.1.\n",
      "DR   SMR; Q91G88; -.\n",
      "DR   KEGG; vg:1733056; -.\n",
      "DR   OrthoDB; 8312at10239; -.\n",
      "DR   Proteomes; UP000001359; Genome.\n",
      "DR   InterPro; IPR022549; DUF3627.\n",
      "DR   InterPro; IPR018004; KilA/APSES_HTH.\n",
      "DR   InterPro; IPR017880; KilA_N.\n",
      "DR   Pfam; PF12299; DUF3627; 1.\n",
      "DR   Pfam; PF04383; KilA-N; 1.\n",
      "DR   PROSITE; PS51301; KILA_N; 1.\n",
      "PE   3: Inferred from homology;\n",
      "KW   Coiled coil; Reference proteome.\n",
      "FT   CHAIN           1..352\n",
      "FT                   /note=\"Putative KilA-N domain-containing protein 006L\"\n",
      "FT                   /id=\"PRO_0000377958\"\n",
      "FT   DOMAIN          15..123\n",
      "FT                   /note=\"KilA-N\"\n",
      "FT                   /evidence=\"ECO:0000255|PROSITE-ProRule:PRU00631\"\n",
      "FT   COILED          129..236\n",
      "FT                   /evidence=\"ECO:0000255\"\n",
      "SQ   SEQUENCE   352 AA;  41658 MW;  3519C5E9E45A8AB6 CRC64;\n",
      "MDSLNEVCYE QIKGTFYKGL FGDFPLIVDK KTGCFNATKL CVLGGKRFVD WNKTLRSKKL\n",
      "IQYYETRCDI KTESLLYEIK GDNNDEITKQ ITGTYLPKEF ILDIASWISV EFYDKCNNII\n",
      "INYFVNEYKT MDKKTLQSKI NEVEEKMQKL LNEKEEELQE KNDKIDELIL FSKRMEEDRK\n",
      "KDREMMIKQE KMLRELGIHL EDVSSQNNEL IEKVDEQVEQ NAVLNFKIDN IQNKLEIAVE\n",
      "DRAPQPKQNL KRERFILLKR NDDYYPYYTI RAQDINARSA LKRQKNLYNE VSVLLDLTCH\n",
      "PNSKTLYVRV KDELKQKGVV FNLCKVSISN SKINEEELIK AMETINDEKR DV\n",
      "//\n",
      "ID   006R_FRG3G              Reviewed;          75 AA.\n",
      "AC   Q6GZW9;\n",
      "DT   28-JUN-2011, integrated into UniProtKB/Swiss-Prot.\n",
      "DT   19-JUL-2004, sequence version 1.\n",
      "DT   09-APR-2025, entry version 36.\n",
      "DE   RecName: Full=Uncharacterized protein 006R;\n",
      "GN   ORFNames=FV3-006R;\n",
      "OS   Frog virus 3 (isolate Goorha) (FV-3).\n",
      "OC   Viruses; Varidnaviria; Bamfordvirae; Nucleocytoviricota; Megaviricetes;\n",
      "OC   Pimascovirales; Iridoviridae; Alphairidovirinae; Ranavirus; Frog virus 3.\n",
      "OX   NCBI_TaxID=654924;\n",
      "OH   NCBI_TaxID=30343; Dryophytes versicolor (chameleon treefrog).\n",
      "OH   NCBI_TaxID=8404; Lithobates pipiens (Northern leopard frog) (Rana pipiens).\n",
      "OH   NCBI_TaxID=45438; Lithobates sylvaticus (Wood frog) (Rana sylvatica).\n",
      "OH   NCBI_TaxID=8316; Notophthalmus viridescens (Eastern newt) (Triturus viridescens).\n",
      "RN   [1]\n",
      "RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].\n",
      "RX   PubMed=15165820; DOI=10.1016/j.virol.2004.02.019;\n",
      "RA   Tan W.G., Barkman T.J., Gregory Chinchar V., Essani K.;\n",
      "RT   \"Comparative genomic analyses of frog virus 3, type species of the genus\n",
      "RT   Ranavirus (family Iridoviridae).\";\n",
      "RL   Virology 323:70-84(2004).\n",
      "CC   ---------------------------------------------------------------------------\n",
      "CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms\n",
      "CC   Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "CC   ---------------------------------------------------------------------------\n",
      "DR   EMBL; AY548484; AAT09665.1; -; Genomic_DNA.\n",
      "DR   RefSeq; YP_031584.1; NC_005946.1.\n",
      "DR   KEGG; vg:2947778; -.\n",
      "DR   Proteomes; UP000008770; Segment.\n",
      "PE   4: Predicted;\n",
      "KW   Reference proteome.\n",
      "FT   CHAIN           1..75\n"
     ]
    }
   ],
   "source": [
    "# Let's see what we're working with.\n",
    "\n",
    "with open('uniprot_sprot.dat') as f_in:\n",
    "    for l_i, line in enumerate(f_in):\n",
    "        print(line.strip())\n",
    "\n",
    "        if l_i > 500:\n",
    "            break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "lEF2uJmwNzJl"
   },
   "source": [
    "### Specify UniprotEntry class\n",
    "Stores information about uniprot entry sequence and features\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "7Eg93s2t-63H"
   },
   "outputs": [],
   "source": [
    "def feature_is_header(line_data):\n",
    "    c1_check = line_data[0] == 'FT'\n",
    "    c2_check = re.match(r\"[A-Z]+\", line_data[1])\n",
    "    c3_check = re.match(r\"<?\\d+(?:\\.\\.\\>?d+)?\", line_data[2])\n",
    "    return c1_check and c2_check and c3_check\n",
    "\n",
    "def get_pos(pos_data):\n",
    "    pos_data = pos_data.replace('<', '')\n",
    "    pos_data = pos_data.replace('>', '')\n",
    "\n",
    "    try:\n",
    "        pos_data = [int(x)-1 for x in pos_data.split('..')]\n",
    "    except ValueError:\n",
    "        return None\n",
    "\n",
    "    if len(pos_data) == 1:\n",
    "        pos_data.append(pos_data[0]+1)\n",
    "\n",
    "    return pos_data\n",
    "\n",
    "def get_subfeature(feature_string):\n",
    "    feature_search = re.search(r'^\\/(.+)=\"(.+)\"', feature_string)\n",
    "    if feature_search:\n",
    "        return feature_search.groups()\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "class UniprotEntry:\n",
    "    def __init__(self, entry_ID):\n",
    "        self.entry_ID = entry_ID\n",
    "        self.features = []\n",
    "        self.sequence = ''\n",
    "\n",
    "    def handle_FT(self, line):\n",
    "        line_data = (line[:2], line[2:21].strip(), line[21:].strip())\n",
    "\n",
    "        if feature_is_header(line_data):\n",
    "            if len(self.features) > 0:\n",
    "                if self.features[-1] == 'ERR':\n",
    "                    self.features.pop()\n",
    "\n",
    "            feature_pos = get_pos(line_data[2])\n",
    "            if feature_pos:\n",
    "                new_feature = [line_data[1], feature_pos, {}]\n",
    "                self.features.append(new_feature)\n",
    "            else:\n",
    "                self.features.append('ERR')\n",
    "        else:\n",
    "            if self.features[-1] == 'ERR':\n",
    "                return\n",
    "\n",
    "            subfeature = get_subfeature(line_data[2])\n",
    "\n",
    "            if subfeature:\n",
    "                if subfeature[0] not in ['evidence', 'id']:\n",
    "                    self.features[-1][-1][subfeature[0]] = subfeature[1]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "NGmyydJlOuYS"
   },
   "source": [
    "### Process Uniprot Data\n",
    "Iterate through `uniprot_sprot.dat` and collect all features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "l6ZvK6Z6q5Ay"
   },
   "outputs": [],
   "source": [
    "with open('uniprot_sprot.dat') as uniprot_in:\n",
    "    uniprot_data = []\n",
    "\n",
    "    for l_i, line in enumerate(uniprot_in):\n",
    "        # We are doing this for demo purposes\n",
    "        # so cap the number of entries at 100\n",
    "        if len(uniprot_data) > 100:\n",
    "            break\n",
    "            \n",
    "        c1 = line[:2]\n",
    "\n",
    "        if c1 == 'ID':\n",
    "            if len(uniprot_data) > 0:\n",
    "                if uniprot_data[-1].features[-1] == 'ERR':\n",
    "                    uniprot_data[-1].features.pop()\n",
    "\n",
    "            entry_ID = re.split(r'\\s+', line)[1]\n",
    "            uniprot_data.append(UniprotEntry(entry_ID))\n",
    "\n",
    "        if c1 == 'FT':\n",
    "            uniprot_data[-1].handle_FT(line)\n",
    "\n",
    "        if c1 == '  ':\n",
    "            seq_data = re.split(r'\\s+', line)\n",
    "            uniprot_data[-1].sequence += ''.join(seq_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "id": "LXLWbve9PRD3",
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "001R_FRG3G\n",
      "MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQVECPKAPVEWNNPPSEKGLIVGHFSGIKYKGEKAQASEVDVNKMCCWVSKFKDAMRRYQGIQTCKIPGKVLSDLDAKIKAYNLTVEGVEGFVRYSRVTKQHVAAFLKELRHSKQYENVNLIHYILTDKRVDIQHLEKDLVKDFKALVESAHRMRQGHMINVKYILYQLLKKHGHGPDGPDILTVKTGSKGVLYDDSFRKIYTDLGWKFTPL\n",
      "[['CHAIN', [0, 255], {'note': 'Putative transcription factor 001R'}]]\n",
      "-----\n",
      "002L_FRG3G\n",
      "MSIIGATRLQNDKSDTYSAGPCYAGGCSAFTPRGTCGKDWDLGEQTCASGFCTSQPLCARIKKTQVCGLRYSSKGKDPLVSAEWDSRGAPYVRCTYDADLIDTQAQVDQFVSMFGESPSLAERYCMRGVKNTAGELVSRVSSDADPAGGWCRKWYSAHRGPDQDAALGSFCIKNPGAADCKCINRASDPVYQKVKTLHAYPDQCWYVPCAADVGELKMGTQRDTPTNCPTQVCQIVFNMLDDGSVTMDDVKNTINCDFSKYVPPPPPPKPTPPTPPTPPTPPTPPTPPTPPTPRPVHNRKVMFFVAGAVLVAILISTVRW\n",
      "[['CHAIN', [0, 319], {'note': 'Uncharacterized protein 002L'}], ['TRANSMEM', [300, 317], {'note': 'Helical'}], ['REGION', [260, 293], {'note': 'Disordered'}], ['COMPBIAS', [261, 293], {'note': 'Pro residues'}]]\n",
      "-----\n",
      "002R_IIV3\n",
      "MASNTVSAQGGSNRPVRDFSNIQDVAQFLLFDPIWNEQPGSIVPWKMNREQALAERYPELQTSEPSEDYSGPVESLELLPLEIKLDIMQYLSWEQISWCKHPWLWTRWYKDNVVRVSAITFEDFQREYAFPEKIQEIHFTDTRAEEIKAILETTPNVTRLVIRRIDDMNYNTHGDLGLDDLEFLTHLMVEDACGFTDFWAPSLTHLTIKNLDMHPRWFGPVMDGIKSMQSTLKYLYIFETYGVNKPFVQWCTDNIETFYCTNSYRYENVPRPIYVWVLFQEDEWHGYRVEDNKFHRRYMYSTILHKRDTDWVENNPLKTPAQVEMYKFLLRISQLNRDGTGYESDSDPENEHFDDESFSSGEEDSSDEDDPTWAPDSDDSDWETETEEEPSVAARILEKGKLTITNLMKSLGFKPKPKKIQSIDRYFCSLDSNYNSEDEDFEYDSDSEDDDSDSEDDC\n",
      "[['CHAIN', [0, 457], {'note': 'Uncharacterized protein 002R'}], ['REGION', [338, 396], {'note': 'Disordered'}], ['REGION', [433, 457], {'note': 'Disordered'}], ['COMPBIAS', [343, 389], {'note': 'Acidic residues'}], ['COMPBIAS', [435, 457], {'note': 'Acidic residues'}]]\n",
      "-----\n",
      "003L_IIV3\n",
      "MYQAINPCPQSWYGSPQLEREIVCKMSGAPHYPNYYPVHPNALGGAWFDTSLNARSLTTTPSLTTCTPPSLAACTPPTSLGMVDSPPHINPPRRIGTLCFDFGSAKSPQRCECVASDRPSTTSNTAPDTYRLLITNSKTRKNNYGTCRLEPLTYGI\n",
      "[['CHAIN', [0, 155], {'note': 'Uncharacterized protein 003L'}]]\n",
      "-----\n",
      "003R_FRG3G\n",
      "MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVYQMSNILLTERRQVDRAMGGSDDDGVMVVALSPSDFKTVLGSALLAVERDMVHVVPKYLQTPGILHDMLVLLTPIFGEALSVDMSGATDVMVQQIATAGFVDVDPLHSSVSWKDNVSCPVALLAVSNAVRTMMGQPCQVTLIIDVGTQNILRDLVNLPVEMSGDLQVMAYTKDPLGKVPAVGVSVFDSGSVQKGDAHSVGAPDGLVSFHTHPVSSAVELNYHAGWPSNVDMSSLLTMKNLMHVVVAEEGLWTMARTLSMQRLTKVLTDAEKDVMRAAAFNLFLPLNELRVMGTKDSNNKSLKTYFEVFETFTIGALMKHSGVTPTAFVDRRWLDNTIYHMGFIPWGRDMRFVVEYDLDGTNPFLNTVPTLMSVKRKAKIQEMFDNMVSRMVTS\n",
      "[['SIGNAL', [0, 31], {}], ['CHAIN', [32, 437], {'note': 'Uncharacterized protein 3R'}]]\n",
      "-----\n",
      "004R_FRG3G\n",
      "MNAKYDTDQGVGRMLFLGTIGLAVVVGGLMAYGYYYDGKTPSSGTSFHTASPSFSSRYRY\n",
      "[['CHAIN', [0, 59], {'note': 'Uncharacterized protein 004R'}], ['TRANSMEM', [13, 33], {'note': 'Helical'}], ['REGION', [37, 59], {'note': 'Disordered'}], ['COMPBIAS', [39, 59], {'note': 'Polar residues'}]]\n",
      "-----\n",
      "005L_IIV3\n",
      "MRYTVLIALQGALLLLLLIDDGQGQSPYPYPGMPCNSSRQCGLGTCVHSRCAHCSSDGTLCSPEDPTMVWPCCPESSCQLVVGLPSLVNHYNCLPNQCTDSSQCPGGFGCMTRRSKCELCKADGEACNSPYLDWRKDKECCSGYCHTEARGLEGVCIDPKKIFCTPKNPWQLAPYPPSYHQPTTLRPPTSLYDSWLMSGFLVKSTTAPSTQEEEDDY\n",
      "[['SIGNAL', [0, 23], {}], ['CHAIN', [24, 216], {'note': 'Uncharacterized protein 005L'}]]\n",
      "-----\n",
      "005R_FRG3G\n",
      "MQNPLPEVMSPEHDKRTTTPMSKEANKFIRELDKKPGDLAVVSDFVKRNTGKRLPIGKRSNLYVRICDLSGTIYMGETFILESWEELYLPEPTKMEVLGTLESCCGIPPFPEWIVMVGEDQCVYAYGDEEILLFAYSVKQLVEEGIQETGISYKYPDDISDVDEEVLQQDEEIQKIRKKTREFVDKDAQEFQDFLNSLDASLLS\n",
      "[['CHAIN', [0, 203], {'note': 'Uncharacterized protein 005R'}], ['REGION', [0, 19], {'note': 'Disordered'}]]\n",
      "-----\n",
      "006L_IIV6\n",
      "MDSLNEVCYEQIKGTFYKGLFGDFPLIVDKKTGCFNATKLCVLGGKRFVDWNKTLRSKKLIQYYETRCDIKTESLLYEIKGDNNDEITKQITGTYLPKEFILDIASWISVEFYDKCNNIIINYFVNEYKTMDKKTLQSKINEVEEKMQKLLNEKEEELQEKNDKIDELILFSKRMEEDRKKDREMMIKQEKMLRELGIHLEDVSSQNNELIEKVDEQVEQNAVLNFKIDNIQNKLEIAVEDRAPQPKQNLKRERFILLKRNDDYYPYYTIRAQDINARSALKRQKNLYNEVSVLLDLTCHPNSKTLYVRVKDELKQKGVVFNLCKVSISNSKINEEELIKAMETINDEKRDV\n",
      "[['CHAIN', [0, 351], {'note': 'Putative KilA-N domain-containing protein 006L'}], ['DOMAIN', [14, 122], {'note': 'KilA-N'}], ['COILED', [128, 235], {}]]\n",
      "-----\n",
      "006R_FRG3G\n",
      "MYKMYFLKDQKFSLSGTIRINDKTQSEYGSVWCPGLSITGLHHDAIDHNMFEEMETEIIEYLGPWVQAEYRRIKG\n",
      "[['CHAIN', [0, 74], {'note': 'Uncharacterized protein 006R'}]]\n",
      "-----\n",
      "007R_FRG3G\n",
      "MRSIKPLRCCNAHGRHVSQEYGRCTLLLFREKLFLQTGLVCNKQCNAPNNDGAESKHHGIHHGSRGALALRGAGVHLLASAALGPRVLAGLVPTGRSVQGSVGQCGRVAQIGRARDVAARKQESYCEK\n",
      "[['CHAIN', [0, 127], {'note': 'Uncharacterized protein 007R'}]]\n",
      "-----\n",
      "007R_IIV3\n",
      "MEAKNITIDNTTYNFFKFYNINQPLTNLKYLNSERLCFSNAVMGKIVDDASTITITYHRVYFGISGPKPRQVADLGEYYDVNELLNYDTYTKTQEFAQKYNSLVKPTIDAKNWSGNELVLLVGNEWYCKTFGKAGSKNVFLYNMIPTIYRDEPQHQEQILKKFMFFNATKNVEQNPNFLDNVPEEYYHLLLPKSWVEKNLSDKYRKIMETEHKPLVFSCEPAFSFGLCRNTQDKNESYQLSLCLYEREKPRDAEIVWAAKYDELAAMVRDYLKKTPEFKKYRSFISCMKGLSWKNNEIGDKDGPKLYPKVIFNRKKGEFVTIFTKDDDVEPETIEDPRTILDRRCVVQAALRLESVFVHNKVAIQLRINDVLISEWKEASSKPQPLILRRHRFTKPSSSVAKSTSPSLRNSGSDESDLNQSDSDKEDERVVPVPKTKRIVKTVKLPN\n",
      "[['CHAIN', [0, 446], {'note': 'Uncharacterized protein 007R'}], ['REGION', [391, 434], {'note': 'Disordered'}], ['COMPBIAS', [394, 406], {'note': 'Low complexity'}], ['COMPBIAS', [407, 420], {'note': 'Polar residues'}]]\n",
      "-----\n",
      "008L_IIV3\n",
      "MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQNAGDVTNKAYVDQAVMSAAVPVASSTTVGTIQMAGDLEGSSGTNPIIAANKITLNKLQKIGPKMVIGNPNSDWNNTQEIELDSSFRIVDNRLNAGIVPISSTDPNKSNTVIPAPQQNGLFYLDSSGRVWVWAEHYYKCITPSRYISKWMGVGDFQELTVGQSVMWDSGRPSIETVSTQGLEVEWISSTNFTLSSLYLIPIVVKVTICIPLLGQPDQMAKFVLYSVSSAQQPRTGIVLTTDSSRSSAPIVSEYITVNWFEPKSYSVQLKEVNSDSGTTVTICSDKWLANPFLDCWITIEEVG\n",
      "[['CHAIN', [0, 346], {'note': 'Uncharacterized protein 008L'}]]\n",
      "-----\n",
      "009L_FRG3G\n",
      "MDTSPYDFLKLYPWLSRGEADKGTLLDAFPGETFEQSLASDVAMRRAVQDDPAFGHQKLVETFLSEDTPYRELLLFHAPGTGKTCTVVSVAERAKEKGLTRGCIVLARGAALLRNFLHELVFNCGTGGRYIPEGYADMGDQERTRKMRKAVSSYYQFRTYETFAKSVATMSAEAIRARYDRFVIVMDEVHHLRSVQAEGVNTYSAISRFLRTVRGCVKMLLTGTPMTNEPGELADVLNLILPQDKTIRPEDGIFSNSGDLLKPDELAERVRGRVSYLKAARPDAGLTFAGEVLGGTGMTHLRLVRLEMSAFQSDAYASAWDQDAGDRNIFSNSRQCSLAVMPDRRWGSAAEARNPSQVRRMAGQNLAEYSVKYDYLVRVASSSPKTFAYCEYVNGSGLSLLSDILLANGWRRATGRETTPGKRFALLTASQKNIHKIVQRFNHEDNVDGAYISLLLGSRVVAEGLTFKEVRHTVILTPHWNYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPRSIDSDMYAVSEVKDKRIKAVERILMTSAADCSLLRSRNLYPSEFDGSRECEYGRCAYRCSNVSVEPGPLPALLGASAAEAVAQVRLDGGGDPAIMKVDMSTLWAEVTAGRRYVNRWGDGAVLRAEGGRLELSAPYGSSEEGRWGDFYKTRNLCYAKMDQDHLRADDLRDSLPQEVEELLTVSPVETIGETASAMPQEVATAILMACVQARADGKTLNVVRRDALLDFYKGFYAMGPSGWTVWLHARGANAKVYDGRRWNPADEDTLEFLAARSAKFTDTRIGYYGLYNPNLKDFCIRDVTQGKRDKVDLRKLTVGRRCVDWDQRTLVHIVARLMKIDGRRDFMPHATLREMRELAEQDPLHEPSDLTSKEACRRFLFWTQKGDNKFRRQDICKAMEKWFIENDLMEDNFDCGHQHKRRGKFA\n",
      "[['CHAIN', [0, 947], {'note': 'Putative helicase 009L'}], ['DOMAIN', [63, 242], {'note': 'Helicase ATP-binding'}], ['DOMAIN', [370, 553], {'note': 'Helicase C-terminal'}], ['MOTIF', [186, 189], {'note': 'DEAH box'}], ['BINDING', [76, 83], {'ligand': 'ATP', 'ligand_id': 'ChEBI:CHEBI:30616'}]]\n",
      "-----\n",
      "009R_IIV6\n",
      "MIKLFCVLAAFISINSACQSSHQQREEFTVATYHSSSICTTYCYSNCVVASQHKGLNVESYTCDKPDPYGRETVCKCTLIKCHDI\n",
      "[['SIGNAL', [0, 19], {}], ['CHAIN', [20, 84], {'note': 'Uncharacterized protein 009R'}]]\n",
      "-----\n",
      "010R_FRG3G\n",
      "MKMDTDCRHWIVLASVPVLTVLAFKGEGALALAGLLVMAAVAMYRDRTEKKYSAARAPSPIAGHKTAYVTDPSAFAAGTVPVYPAPSNMGSDRFEGWVGGVLTGVGSSHLDHRKFAERQLVDRREKMVGYGWTKSFF\n",
      "[['CHAIN', [0, 136], {'note': 'Uncharacterized protein 010R'}], ['TRANSMEM', [19, 41], {'note': 'Helical'}]]\n",
      "-----\n",
      "011L_IIV3\n",
      "MMESPKYKKSTCSVTNLGGTCILPQKGATAPKAKDVSPELLVNKMDNLCQDWARTRNEYNKVHIEQAPTDSYFGVVHSHTPKKKYTSRDSDSEPEATSTRRSATAQRAANLKSSPVDQWSTTPPQPQPQPAAPTVKKTCASSPPAALSVKRTCTSPPPPPVLIDDDTGEDAFYDTNDPDIFYDIENGVSELETEGPKRPVYYQRNIRYPIDGSVPQESEQWYDPIDDEFLASSGDVVSLEPSPIAAFQPTPPKTVQFVPMPEEIIVPPPPPPKTVVDEGVQAMPYTVDQMIQTDFEESPLLANVNLRTIPIEEVNPNFSPVLMQDMVRDSFVFGTVAQRVMASQRVKQFFKELIEQDVSLAGRMCMDSGSPQLNLYNSLMGVKLLYRWRSSTTFYRAIVPEIDEPVQVMQDVLSSSEWAKFDSQAGIPPKMVYIHYKLLNDLVKTLICPNFQLTHAALVCVDCRPEAVGSDGLQDGRQRRCSNLVSEYHEMTLEDLFNTIKPADLNAKNIILSVLFQMLYAVATVQKQFGMGGLFANADSVHVRRIQPGGFWHYTVNGLRYSVPNYGYLVILTNFTDVVNYRPDFATTRYFGRRQAKVVPTRNWYKFVPFTTRYRPFVTVDPITQAKTTAYAPNPPTEGITINEFYKDSSDLRPSVPVDLNDMITFPVPEFHLTICRLFSFFSKFYDSNFIGNDPFVRNLVDRYSQPFEFPDVYWPEDGVSRVLACYTIEEIYPNWVDGDTDYVIESYNLD\n",
      "[['CHAIN', [0, 750], {'note': 'Uncharacterized protein 011L'}], ['REGION', [72, 168], {'note': 'Disordered'}], ['COMPBIAS', [95, 108], {'note': 'Low complexity'}], ['COMPBIAS', [110, 119], {'note': 'Polar residues'}]]\n",
      "-----\n",
      "011R_FRG3G\n",
      "MTSVKTIAMLAMLVIVAALIYMGYRTFTSMQSKLNELESRVNAPQLRPPVMSPIVPLNFIESEDLDKELD\n",
      "[['CHAIN', [0, 69], {'note': 'Uncharacterized protein 011R'}], ['TRANSMEM', [3, 23], {'note': 'Helical'}]]\n",
      "-----\n",
      "012L_FRG3G\n",
      "MCAKLVEMAFGPVNADSPPLTAEEKESAVEKLVGSKPFPALKKKYHDKVPAQDPKYCLFSFVEVLPSCDIKAAGAEEMCSCCIKRRRGQVFGVACVRGTAHTLAKAKQKADKLVGDYDSVHVVQTCHVGRPFPLVSSGMAQETVAPSAMEAAEAAMDAKSAEKRKERMRQKLEMRKREQEIKARNRKLLEDPSCDPDAEEETDLERYATLRVKTTCLLENAKNASAQIKEYLASMRKSAEAVVAMEAADPTLVENYPGLIRDSRAKMGVSKQDTEAFLKMSSFDCLTAASELETMGF\n",
      "[['CHAIN', [0, 296], {'note': 'Uncharacterized protein 012L'}]]\n",
      "-----\n",
      "013L_IIV3\n",
      "MYYRDQYGNVKYAPEGMGPHHAASSSHHSAQHHHMTKENFSMDDVHSWFEKYKMWFLYALILALIFGVFMWWSKYNHDKKRSLNTASIFY\n",
      "[['CHAIN', [0, 89], {'note': 'Uncharacterized protein IIV3-013L'}], ['TRANSMEM', [51, 71], {'note': 'Helical'}], ['REGION', [12, 33], {'note': 'Disordered'}]]\n",
      "-----\n",
      "013R_FRG3G\n",
      "MANSVAFSSMTWYSPLASDNLYDICVDKVHNRVLCLCHSFGCCTNAVVIWILPSFDEFTPQTLSCKGP\n",
      "[['CHAIN', [0, 67], {'note': 'Uncharacterized protein 013R'}]]\n",
      "-----\n",
      "014R_FRG3G\n",
      "METLVQAYLDIQGKIAEFRREIKALRVEEKAITANLFEAMGEAGVESIRISEDRYLVAEEKPKRTRSKQQFYQAAEGEGFTQEDVDRLMSLSRGAVTGSSSNVKIRKSAPARNEEDDDG\n",
      "[['CHAIN', [0, 118], {'note': 'Uncharacterized protein 014R'}], ['REGION', [94, 118], {'note': 'Disordered'}], ['COILED', [5, 35], {}]]\n",
      "-----\n",
      "015R_FRG3G\n",
      "MEQVPIKEMRLSDLRPNNKSIDTDLGGTKLVVIGKPGSGKSTLIKALLDSKRHIIPCAVVISGSEEANGFYKGVVPDLFIYHQFSPSIIDRIHRRQVKAKAEMGSKKSWLLVVIDDCMDNAKMFNDKEVRALFKNGRHWNVLVVIANQYVMDLTPDLRSSVDGVFLFRENNVTYRDKTYANFASVVPKKLYPTVMETVCQNYRCMFIDNTKATDNWHDSVFWYKAPYSKSAVAPFGARSYWKYACSKTGEEMPAVFDNVKILGDLLLKELPEAGEALVTYGGKDGPSDNEDGPSDDEDGPSDDEEGLSKDGVSEYYQSDLDD\n",
      "[['CHAIN', [0, 321], {'note': 'Uncharacterized protein 015R'}], ['REGION', [276, 321], {'note': 'Disordered'}], ['COMPBIAS', [286, 304], {'note': 'Acidic residues'}]]\n",
      "-----\n",
      "017L_FRG3G\n",
      "METMSDYSKEVSEALSALRGELSALSAAISNTVRAGSYSAPVAKDCKAGHCDSKAVLKSLSRSARDLDSAVEAVSSNCEWASSGYGKQIARALRDDAVRVKREVESTRDAVDVVTPSCCVQGLAEEAGKLSEMAAVYRCMATVFETADSHGVREMLAKVDGLKQTMSGFKRLLGKTAEIDGLSDSVIRLGRSIGEVLPATEGKAMRDLVKQCERLNGLVVDGSRKVEEQCSKLRDMASQSYVVADLASQYDVLGGKAQEALSASDALEQAAAVALRAKAAADAVAKSLDSLDVKKLDRLLEQASAVSGLLAKKNDLDAVVTSLAGLEALVAKKDELYKICAAVNSVDKSKLELLNVKPDRLKSLTEQTVVVSQMTTALATFNEDKLDSVLGKYMQMHRFLGMATQLKLMSDSLAEFQPAKMAQMAAAASQLKDFLTDQTVSRLEKVSAAVDATDVTKYASAFSDGGMVSDMTKAYETVKAFAAVVNSLDSKKLKLVAECAKK\n",
      "[['CHAIN', [0, 501], {'note': 'Uncharacterized protein 017L'}]]\n",
      "-----\n",
      "018L_FRG3G\n",
      "MQNSKTDMCAALWAVTGLVLNVAVRFALEPFKESMGQGWHTAARVAVNGAIVLALADRLSDSPVTMTLFVMALSASPE\n",
      "[['SIGNAL', [0, 26], {}], ['CHAIN', [27, 77], {'note': 'Uncharacterized protein 018L'}]]\n",
      "-----\n",
      "019R_FRG3G\n",
      "MATNYCDEFERNPTRNPRTGRTIKRGGPVFRALERECSDGAARVFPAAAVRGAAAARAASPRVAAASPCPEFARDPTRNPRTGRPIKRGGPVFRALERECADYGGASPRRVSPARAFPNRRVSPARRQSPAEAAEASPCPEFARDPTRNPRTGRTIKRGGPTYRALEAECADYGRLSPIRSPWSDWSSTGLSPFRSHMRKSPARRSPARRSPARRSLARYTEHLTSDSETEVDYDARNVIRSQVGPGGVCERFAADPTRNPVTGSPLSRNDPLYTDLMEICKGYPDTPLTKSLTGEGTDDDTCEAFCRDPTRNPVTGQKMRRNGIEYQMFAEECDCSGISRPSGVSRTSGTSGSSGSSASSRPPNSFEAPGASSRPPNSFEASGAARVPGTPSVSRGEPRWMSSISTRHNYDESNPMSVAFRLRHVKDIRKFLRTVRPGRSGFCATDKGGWLGSAAVSDNVIGQGSWGSVHMVKFRDFPEEFVVKEAVLMSVSEKHRYKPTVVWDEWAAGSVPDEVVVNNMVTEIAATGMTPFVPLTAGAGACDSCNPQLLEKAAKVTKCYLQAMEAADFSLDRVLPTMSPDQAASALAQILLGLQSLQTTLGIMHNDIKAHNILVKRVPPGGYWKVTDSFNGQVFYIPNEGYLCMLADYGVVRLVKPAVGMDTLYGTRNARFVPRDVGRWGKGAGTEYVVTPIRSKISVVVRGGRFVGVEPNKAVRYWKNTDTSKVGDVITTNNVFYMGYDIEPDMQVQLDDTNSFPVWESRGDVADCVRTFVGGKRASQPGFHRLFYKKTGSAWEKAAETVAKQNPLFSGFTLDGSGLKYIRAATACAYIFPGMAVPRPGEREIESFTM\n",
      "[['CHAIN', [0, 850], {'note': 'Putative serine/threonine-protein kinase 019R'}], ['DOMAIN', [455, 850], {'note': 'Protein kinase'}], ['REGION', [0, 23], {'note': 'Disordered'}], ['REGION', [60, 90], {'note': 'Disordered'}], ['REGION', [103, 159], {'note': 'Disordered'}], ['REGION', [189, 215], {'note': 'Disordered'}], ['REGION', [339, 399], {'note': 'Disordered'}], ['COMPBIAS', [124, 140], {'note': 'Low complexity'}], ['COMPBIAS', [195, 215], {'note': 'Basic residues'}], ['COMPBIAS', [339, 365], {'note': 'Low complexity'}], ['ACT_SITE', [607, 608], {'note': 'Proton acceptor'}], ['BINDING', [461, 469], {'ligand': 'ATP', 'ligand_id': 'ChEBI:CHEBI:30616'}], ['BINDING', [484, 485], {'ligand': 'ATP', 'ligand_id': 'ChEBI:CHEBI:30616'}]]\n",
      "-----\n",
      "020R_FRG3G\n",
      "MLQNYAIVLGMAVAVAIWYFFKIEEEAPPGPNPPKPDPPKPDPPKMHMPKKKPHWMDPHLTGSQTVQYSRNRSMGDPIRGDLPIIPRDDGWFSTAANPAHTLHAGALSMIAPASTGGGLTVNKLISAYADKGNAMSGRHNSPSYYGSS\n",
      "[['CHAIN', [0, 147], {'note': 'Uncharacterized protein 020R'}], ['TRANSMEM', [0, 20], {'note': 'Helical'}], ['REGION', [26, 60], {'note': 'Disordered'}]]\n",
      "-----\n",
      "021L_FRG3G\n",
      "METIVLVPRQDQETFSDSRPVLDGDLMLEFLENKIRHPVRRRQPRVVPVTSSDPEVVDDEDDEDQSDDSDEERQRLYFQYMVLKRMYPTEVIPEMTTYSNVAIMREKYKLLTRRLSLDKHINEWKKYIIVGMCIMELVMTKLNFDASGFARYQIKSLGAYDQLLAEMADKYYEATPQSSVEMRLMTTMGMNMAVFMLGKLLGGQMDFLGLLENAFGSSS\n",
      "[['CHAIN', [0, 218], {'note': 'Uncharacterized protein 021L'}], ['REGION', [41, 70], {'note': 'Disordered'}], ['COMPBIAS', [44, 53], {'note': 'Low complexity'}], ['COMPBIAS', [54, 70], {'note': 'Acidic residues'}]]\n",
      "-----\n",
      "022L_IIV3\n",
      "MSFVHKLPTFYTAGVGAIIGGLSLRFNGAKFLSDWYINKYNDSVPAWSLQTCHWAGIALYCVGWVTLASVIYLKHRDNSILKGSILSCIVISAVWSILEYNQDMFVSNPKLPLISCAMLVSSLAALVALKYHIKDIFTILGAAIIIILAEYVVLPYQRQYNIVDGIGLPLLLLGFFILYQVFSVPNPSTPTGVMVPKPEDEWDIEMAPLNHRDRQVPESELENVK\n",
      "[['CHAIN', [0, 224], {'note': 'Transmembrane protein 022L'}], ['TRANSMEM', [1, 21], {'note': 'Helical'}], ['TRANSMEM', [52, 72], {'note': 'Helical'}], ['TRANSMEM', [79, 99], {'note': 'Helical'}], ['TRANSMEM', [110, 130], {'note': 'Helical'}], ['TRANSMEM', [135, 155], {'note': 'Helical'}], ['TRANSMEM', [161, 181], {'note': 'Helical'}]]\n",
      "-----\n",
      "023R_FRG3G\n",
      "MRVSQTSWIVSRMLEYPRGGFFYSTDMACMMEGLAEELAGGHKDEVLIVSGRNGDDEVFKEFPNVRAADGLKGPNSIDPETKLVLIIDVSPTAISNALAATLQEFLIPVWVFCNHTRTLTASVTRRLGYKLWPKGTYTPYICEKAGVSEVVTYNQPESEKFVAFMSAARQIMDKRKSKKTMQELAFLPHLAFAEIAMEGDQEMTPTLTAKKVSDIKDEQVNELASAMFRTGKLSHLDMLSVPDCVYSCGEALKREVAKAKANRERFVVALRNAQYKKYTAGLLEAGTPVKTFTEVIKNWGAYDTIFLPMGVDWTYTGGSNLIRMMMTPGSHKTVTFVPESDDVHEFCHNKPTVNTMGVESAATGLAAELNRRWRRDNPVDAS\n",
      "[['CHAIN', [0, 381], {'note': 'Uncharacterized protein 023R'}]]\n",
      "-----\n",
      "023R_IIV3\n",
      "MGSYMLFDSLIKLVENRNPLNHEQKLWLIDVINNTLNLEGKEKLYSLLIVHNKQQTKIYDPKEPFYDIEKIPVQLQLVWYEFTKMHLKSQNEDRRRKMSLYAGRSP\n",
      "[['CHAIN', [0, 105], {'note': 'Uncharacterized protein 023R'}]]\n",
      "-----\n",
      "024R_FRG3G\n",
      "MWQYLPILLMTMISQLEWTVAAVKRYPAGGFITGDKLSRVFEALPWRVAVVSDEPEKYEGFPILTEEDPAVFEDADCILFAVSDPKCVTGAMKSVFMASSKTAWVVYDGTETRATVRSWMRRLWRAETYVPLLTHRGFVTDVCVYSQPDSERYVSVMTATAHFYSNRLEVLEEMAFVPHLAYAKLAMGRYTVLDGCMSVKGSADVAPLNRSMWFLTAAAIPHGEIDTDSLFSDPGAVYSCGSALREALGSLPEGSTSVVAVRNSSYRKYVRGILGPNFRVETFTNVVKTWGVYDYVLLPMGISDSYKQGRDLMEKLEMPGGHRVVTFAPENYTVNEVHLNRPLKYAIKRMDLITPMVLRHVSLNK\n",
      "[['SIGNAL', [0, 20], {}], ['CHAIN', [21, 364], {'note': 'Uncharacterized protein 024R'}]]\n",
      "-----\n",
      "025R_IIV3\n",
      "MNYSVIWAITILILGLVLTLAWARQNPTHPINPLVLNYHTKPSPKRHRMVLVVESFASVDALVELVENILSQTIRVASITVVSQRPDHLRQVPLLHQTCTFSRASGLSALFKETSGTLVVFISKEGFHHFQSPTLLETIDQRGVTAEQTLPGIVLRNTDMPGIDLTTVYRQQRLGLGN\n",
      "[['SIGNAL', [0, 22], {}], ['CHAIN', [23, 177], {'note': 'Uncharacterized protein 025R'}]]\n",
      "-----\n",
      "026R_IIV6\n",
      "MAISFFSDTSYIIKSILLISLFSIIPLEDEVTKLKSSSLRETSELNKEEGITTCLYTFN\n",
      "[['CHAIN', [0, 58], {'note': 'Uncharacterized protein 026R'}]]\n",
      "-----\n",
      "027R_FRG3G\n",
      "MANFLQDVNCETVSEYDGPDASIPEGVWEGYVGHDHAALWRTWSYIYECCKKGTLVQFRGGKLVTFSMFDNPRFSNGAGIDAQKVLDLEDRARELQGYGPVNRRTDVMPVDRWTLNGPLLRYDKMVLEDVGGTGSNRTMVRAQLEALQDERDVPDCDFILNVRDYPLLRRDGTRPYPQVYGKGRRLPEPWARGGPHVPVVSMCSGPTYADIAVPTYECIAHAYTSSGRTLPAGGRFVKTPSADSLPAWRDRKALAVFRGSSTGAGTSTEDNQRLRALQISMSRPDLADVGITKWNLRPRKTERYDGYRIIEPWQFGRKSPYPAAAKPMTPEQIAGYKYVLCLWGHAPAFRLARDLSLGSVVLLPSRPPGQEGLDMWHSSVLKPWTHYIPVRGDLSDLEKRIEWCRDNDAECEKIAAAGMEASLNLLGWEGQLDRWMDVLRSVRLECCPGGYDMPPSPSLVSDSMCVRQMVSFPRYEDIPQPSSPMPVLPRCSGTLRGWGLAASLGWDLGDAAEVLNVKRSTAVLSKTVFNNLIYRTPHLRYTFGVAASDPESTAAVILSEKLKGAVTMRSWLEDSRAWARGRNVASVLCQVSQALLEAQAAAGTVFGDLSLDTILVVPNPLPEYIYHDGTGGSFGLKLMPGDKWAVVTYGDYTRARIRVLKGDGRKGHLAVVGPQPVYTKLSERKWHDICCLVSCILRTARTSKRPAARALAAAVARAAGVKRPDMDAEALEATPYEAREEPLTRFGPAEFINGLVREFKLEEGGWAWTEKNKNIEKVLRPWERGLPLYPVRLWLSGDRKEAMRACVSSVLKAAPPRPATAAGAHHTFQTYLRTVGADLDSFPEWAAAAAHLKRLWKSPGSLPAGSASLRAPSVPPPCHGPAWALPFGTRTPGEFPSWFDPSCLGDWTEAMGQGAPLDLENGPAKAGSDPVAVHSAWETASQLSFEEDGWTESEPRPVRREAHVRAKERH\n",
      "[['CHAIN', [0, 969], {'note': 'Uncharacterized protein 027R'}], ['REGION', [941, 969], {'note': 'Disordered'}], ['COMPBIAS', [952, 969], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "028R_FRG3G\n",
      "MDPNVLKNLSLMLSRRAGVSGGEPPRMIEWPEYGQRSEPCGSQTVWYVDRPVGAPFIKAFASEVEERGGGILIHAGKVTFDSAKKLAAMKEVQVFDVKYFSFDLMAVVPEHSLWKRPGDKGYPEKTAQSFPKIMASDPVCRYHGFRPRDLVHVKPHDVYIVC\n",
      "[['CHAIN', [0, 161], {'note': 'Uncharacterized protein 028R'}]]\n",
      "-----\n",
      "028R_IIV3\n",
      "MDQYITLVELYIYDCNLFKSKNLKSFYKVHRVPEGDIVPKRRGGQLAGVTKSWVETNLVHFPLWLSEWDETRWGVLNHYPLESWLEKNVSSKVPVNPVMWNFDSECLVYFFHNGRRTPFLTPKGVVKLQVFYNLMSGKEVEWFYEISNGFLKPHLHQLSNVRELVRLKHAPVVVGAGGPRLVTEGVYSLRDDDFVVDCSQIAAVKRAIERGESHQSLRKYQCPLFVALTDKFQDTVKLVEKKFEVQLNELKAETTIQVLREQLRQEKKLKEQVLSLTQSFIPTIGGRGEEFGKPDETPSSASVGDDNFPSSTNHTFEARRRPSSLSSGGALKPSKIL\n",
      "[['CHAIN', [0, 336], {'note': 'Uncharacterized protein 028R'}], ['REGION', [284, 336], {'note': 'Disordered'}], ['COILED', [247, 275], {}], ['COMPBIAS', [286, 295], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [296, 314], {'note': 'Polar residues'}]]\n",
      "-----\n",
      "029L_FRG3G\n",
      "MRRMRSGFKHCAIPIDICRWEYILSPLILQDLQGPQQGGSVAVDVTVRCSVRFVHLPHYGGFNHGTVQRRVDPDDCRILRQLHIVLSLRLCLIDRDRL\n",
      "[['CHAIN', [0, 97], {'note': 'Uncharacterized protein 029L'}]]\n",
      "-----\n",
      "029R_IIV6\n",
      "MVERLGIAVEDRSPKLRKQAIRERFVLFKKNTERVEKYEYYAIRGQSIYINGRLSKLQSERYPKMIILLDIFCQPNPRNLFLRFKERIDGKSEWENNFTYAGNNIGCTKEMESDMIRIFNELDDEKRDV\n",
      "[['CHAIN', [0, 128], {'note': 'Uncharacterized protein 029R'}]]\n",
      "-----\n",
      "030L_IIV3\n",
      "MHPTLKSNAGEWSQPIVNLFYSNFSGNCKALLQYIDNAGITDHIPIKFINVDNPTMRSVVSAKISHVPALVVLQDDQMSLYVAESVWEWFDNYRTPPPLADGATVDSQASENGEKEAQPTPPKEGLLTVLELAKQMRKEREQQT\n",
      "[['CHAIN', [0, 143], {'note': 'uncharacterized protein 030L'}], ['REGION', [97, 126], {'note': 'Disordered'}]]\n",
      "-----\n",
      "030R_FRG3G\n",
      "MSLYLLLGLKILRYLKMVIVLRCHSAFLLSVKFLREKRRLKMYLGIMLGF\n",
      "[['CHAIN', [0, 49], {'note': 'Uncharacterized protein 030R'}]]\n",
      "-----\n",
      "031R_FRG3G\n",
      "MDTPCKLFCIELKEGYVPGTVSHNHMMPYFLAGSGWPVEITFHAATVELKTQEDFPPAIGIGIHNMTGVPVVETPHSGRMHFVFIFHSKSGRFSATYKCIPVPVVVRDYKTVASVSLTTLSLEDIVGVKLFGTACDRSS\n",
      "[['CHAIN', [0, 138], {'note': 'Uncharacterized protein 031R'}]]\n",
      "-----\n",
      "032R_FRG3G\n",
      "MVTVTELRATAKNLGIRGYSTMRKAELEEAIRDHGRVSEARVASPRRSPARSPRKSPAGRKSPSKSPAGRKSPSKSPAGRKSPSKSPAGRKSPSKSPAGRKSPSKSPAGRKSPSKSPVRKSPSKSPVRKSPRKSPAAKLQAGDRPASMNICKNLPKQRLVDIATEMGIDLNRESDGKPKTKDQLCADIMGGAGRKSPRKSPSRSPVRKSPSRSPVRKSPVRSPRKSPVRVPSPVRSPVKEKTPVRSPARSEDAGSDLAPRPRRGKAVRLDYDEDDDYSYGASTDNLFSGNKEIPFPTRKRRTRKPEKVFVDVRSPHTLTDSEDEDDMVEVPELEDKEITMPGVLSPYSDEIVERGYVSQGGADYINYIYRTEYALESDESFARGARPKTNKRDSDRAVREAAAAAAIARALDRRSQSGNDEPAVRRRSAPTDSSRESRRDREPQRDIAEPQRDIAEPQRDIAEPQRDIAEPRKVRFREAGSADVRVFERDEPKEYGRVPVRPPLFMPAGEPLQPLKFRPKTPKIDDTIHRAQMVLPSKPSQKETDNYYKQFAGEAVRPSEPVQWDKDDQVLYHKVPAWDDSSYAAAVSAWPMSVDPKQAESVFAEFEQLSAQDSDLIKVRKSIMKALGY\n",
      "[['CHAIN', [0, 628], {'note': 'Uncharacterized protein 032R'}], ['REGION', [29, 154], {'note': 'Disordered'}], ['REGION', [168, 326], {'note': 'Disordered'}], ['REGION', [375, 473], {'note': 'Disordered'}], ['COMPBIAS', [29, 38], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [47, 57], {'note': 'Basic residues'}], ['COMPBIAS', [116, 131], {'note': 'Basic residues'}], ['COMPBIAS', [169, 183], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [194, 226], {'note': 'Basic residues'}], ['COMPBIAS', [236, 251], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [304, 313], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [389, 398], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [399, 408], {'note': 'Low complexity'}], ['COMPBIAS', [432, 473], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "032R_IIV3\n",
      "MKLMLEIVKNISEPVGKLAIWFNETYQVDVSETINKWNELTGMNITVQENAVSADDTTAEETEYSVVVNENPTRTAARTRKESKTAAKPRKMQIPKTKDVCQHIFKSGSRAGEQCTTKPKNNALFCSAHRVRNSVTSNATEASEKTVAKTNGTAAPQKRGVKSKSPTVIPSDFDDSDSSSSATRGLRKAPTLSPRKPPPTTTTASSAQEEEDEQQAHFSGSSSPPPKNNGNGAVYSDSSSDEDDDDAHHTTVIPLLKKGARKPLDENVQFTSDSSDEED\n",
      "[['CHAIN', [0, 278], {'note': 'Uncharacterized protein 032R'}], ['REGION', [135, 278], {'note': 'Disordered'}], ['COMPBIAS', [227, 237], {'note': 'Low complexity'}]]\n",
      "-----\n",
      "032R_IIV6\n",
      "MGVYKFCYNKKKEVGQVAVLQKERLIFYIVTKEKSYLKPTLANFSNAIDSLYNECLLRKCCKLAIPKIGCCLDRLYWKTVKNIIIDKLCKKGIEVVVYYI\n",
      "[['CHAIN', [0, 99], {'note': 'Uncharacterized protein 032R'}]]\n",
      "-----\n",
      "033R_FRG3G\n",
      "MSGIQLDKETILKYSSAALVALSAVVAVMMVSNNSESWKPILVGAVVAASGAAAYQSWWPKQS\n",
      "[['CHAIN', [0, 62], {'note': 'Transmembrane protein 033R'}]]\n",
      "-----\n",
      "034R_FRG3G\n",
      "MSAGHLRKRRYVKVGDIHDMGPILGGVHDVSSPPPNVHYQQQDDHNDPGCMIHYPGEGWFSSMSTVEKLMLGAVIVAAVVVGVRMFMSSGNSSATSSFSTAPYFMG\n",
      "[['CHAIN', [0, 105], {'note': 'Uncharacterized protein 034R'}], ['TRANSMEM', [68, 87], {'note': 'Helical'}]]\n",
      "-----\n",
      "034R_IIV6\n",
      "MKQNLLILLSLLLVVVAIMWWLYEKKKEVPLPPPTPPTPPTPTGVPFLPMYAGLSSPVQYNPADYLYGWEKYPHGPAWSFGDRVPYAEAKNALGGHFGGGLYSPRDPILESKLGGVYIGNDLYTVGGVGGDGHW\n",
      "[['CHAIN', [0, 133], {'note': 'Uncharacterized protein 034R'}], ['TRANSMEM', [3, 23], {'note': 'Helical'}]]\n",
      "-----\n",
      "035L_FRG3G\n",
      "MIWVWPATGRGPGWWGIRRDPWGPEDSSCPCPRLPLSPTGPVGHSGPMGQCHPPVPSYRRGRRDQKDPPLRRQTSPPLPPHPWDRPLPWVPWIPLDLCRHGDPRHPWDPGAQSGYPRVREVRGVPADRPLRPCPRQGPRTAATRKESSCRIPS\n",
      "[['CHAIN', [0, 152], {'note': 'Uncharacterized protein 035L'}], ['REGION', [23, 86], {'note': 'Disordered'}], ['REGION', [100, 152], {'note': 'Disordered'}], ['COMPBIAS', [26, 36], {'note': 'Low complexity'}], ['COMPBIAS', [142, 152], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "036L_FRG3G\n",
      "MTLPDVSGSLGPLSPGTNGTLWAVGPRVVRYQIPALAYLTPGALWTLRTRGTSLTSGPIGTRDSIRTLHAVHYDVWTLGPLGPLGPTSPRGPSARPCRLQTDSLHSTDARCYRCKMLQMQDATDARCKKDMSPFSFPGILEPSHLVGSLKSPRVDPGVPCRPLALWGHPYQCLRLVPLYQRCLHPHCFPAAPGRPWDPWCRPDRLDP\n",
      "[['CHAIN', [0, 206], {'note': 'Uncharacterized protein 036L'}]]\n",
      "-----\n",
      "037L_IIV3\n",
      "MNAATSGIQLNAQTLSQQPAMNTPLIHRSFRDDYTGLVSAGDGLYKRKLKVPSTTRCNKFKWCSIGWSIGALIIFLVYKLEKPHVQPTSNGNLSLIEPEKLVSESQLIQKILNATTPQTTTPEIPSSTEPQELVTEILNTTTPQTTTPEIPSSTEPQELVTEIPSSTEPQEEIFSIFKSPKPEEPGGINSIPQYEQESNNVEDEPPPNKPEEEEDHDNQPLEERHTVPILGDVIIRNKTIIIDGGNETIIIKP\n",
      "[['CHAIN', [0, 252], {'note': 'Uncharacterized protein 037L'}], ['TRANSMEM', [61, 77], {'note': 'Helical'}], ['REGION', [140, 224], {'note': 'Disordered'}], ['COMPBIAS', [140, 157], {'note': 'Low complexity'}], ['COMPBIAS', [199, 215], {'note': 'Acidic residues'}]]\n",
      "-----\n",
      "037R_FRG3G\n",
      "MQVFLDLDETLIHSIPVSRLGWTKSKPYPVKPFTVQDAGTPLSVMMGSSKAVNDGRKRLATRLSLFKRTVLTDHIMCWRPTLRTFLNGLFASGYKINVWTAASKPYALEVVKALNLKSYGMGLLVTAQDYPKGSVKRLKYLTGLDAVKIPLSNTAIVDDREEVKRAQPTRAVHIKPFTASSANTACSESDELKRVTASLAIIAGRSRRR\n",
      "[['CHAIN', [0, 208], {'note': 'uncharacterized protein 037R'}], ['DOMAIN', [0, 198], {'note': 'FCP1 homology'}]]\n",
      "-----\n",
      "039R_FRG3G\n",
      "MTSYCDTLKALAAESDSTGSERATIRMYMAMFSDASLRPAVSDTVASILGTDSLDHEDAEMMLKFKLLFFSGSANASATSHYPKADDPQRFARSVSRGPSRVRRPARNSASRPVRR\n",
      "[['CHAIN', [0, 115], {'note': 'Uncharacterized protein 039R'}], ['REGION', [76, 115], {'note': 'Disordered'}]]\n",
      "-----\n",
      "040R_FRG3G\n",
      "MIRALCTIVLIAAGVAVALYLSLVYGYYMSVGVQDASWLTALTGNRPDAKVPFFDKAVGEAPEDKVAYTERPYPVSSTQSPTTTQSPTTTTLKPTTMAVLASIGATPTPVVCHNVRGDMQGIACNVVMKKTVAAALKVQPEAKKDNVNAQYRYGMWTPLRRSRSPFGVWNIPKKLAIAAPDV\n",
      "[['SIGNAL', [0, 25], {}], ['CHAIN', [26, 181], {'note': 'Uncharacterized protein 040R'}], ['REGION', [67, 89], {'note': 'Disordered'}], ['COMPBIAS', [73, 89], {'note': 'Low complexity'}]]\n",
      "-----\n",
      "040R_IIV3\n",
      "MVTMAIKNFHIQDDRLKNGRGNKTMSESDYNTSDSGGWVLVRKKRDRSTRPPDVVDRWSNSTSTFPMGLDQIKIKRNGCVNTY\n",
      "[['CHAIN', [0, 82], {'note': 'Uncharacterized protein 040R'}], ['REGION', [14, 35], {'note': 'Disordered'}], ['COMPBIAS', [20, 34], {'note': 'Polar residues'}]]\n",
      "-----\n",
      "041L_IIV6\n",
      "MNFIRENETKYVLSTYQSMTPKNLMEYLLKYNYDNDCVYIFNNLPKDLQKEVDDLAKEVVKANDEQIKAQDEQIKANDQKLKQLDVMIEFMKQYNKQLDNDIYLLEHQLENKRELNRQLGIF\n",
      "[['CHAIN', [0, 121], {'note': 'Uncharacterized protein 041L'}], ['COILED', [45, 115], {}]]\n",
      "-----\n",
      "041R_FRG3G\n",
      "MRVVVNAKALEVPVGMSFTEWTRTLSPGSSPRFLAWNPVRPRTFKDVTDPFWNGKVFDLLGVVNGKDDLLFPASEIQEWLEYAPNVDLAELERIFVATHRHRGMMGFAAAVQDSLVHVDPDSVDVTRVKDGLHKELDEHASKAAATDVRLKRLRSVKPVDGFSDPVLIRTVFSVTVPEFGDRTAYEIVDSAVPTGSCPYISAGPFVKTIPGFKPAPEWPAQTAHAEGAVFFKADAEFPDTKPLKDMYRKYSGAAVVPGDVTYPAVITFDVPQGSRHVPPEDFAARVAESLSLDLRGRPLVEMGRVVSVRLDGMRFRPYVLTDLLVSDPDASHVMQTDELNRAHKIKGTVYAQVCGTGQTVSFQEKTDEDSGEAYISLRVRARDRKGVEELMEAAGRVMAIYSRRESEIVSFYALYDKTVAKEAAPPRPPRKSKAPEPTGDKADRKLLRTLAPDIFLPTYSRKCLHMPVILRGAELEDARKKGLNLMDFPLFGESERLTYACKHPQHPYPGLRANLLPNKAKYPFVPCCYSKDQAVRPNSKWTAYTTGNAEARRQGRIREGVMQAEPLPEGALIFLRRVLGQETGSKFFALRTTGVPETPVNAVHVAVFQRSLTAEEQAEERAAMALDPSAMGACAQELYVEPDVDWDRWRREMGDPNVPFNLLKYFRALETRYDCDIYIMDNKGIIHTKAVRGRLRYRSRRPTVILHLREESCVPVMTPPSDWTRGPVRNGILTFSPIDPITVKLHDLYQDSRPVYVDGVRVPPLRSDWLPCSGQVVDRAGKARVFVVTPTGKMSRGSFTLVTWPMPPLAAPILRTDTGFPRGRSDSPLSFLGSRFVPSGYRRSVETGAIREITGILDGACEACLLTHDPVLVPDPSWSDGGPPVYEDPVPSRALEGFTGAEKKARMLVEYAKKAISIREGSCTQESVRSFAANGGFVVSPGALDGMKVFNPRFEAPGPFAEADWAVKVPDVKTARRLVYALRVASVNGTCPVQEYASASLVPNFYKTSTDFVQSPAYTINVWRNDLDQSAVKKTRRAVVDWERGLAVPWPLPETELGFSYSLRFAGISRTFMAMNHPTWESAAFAALTWAKSGYCPGVTSNQIPEGEKVPTYACVKGMKPAKVLESGDGTLKLDKSSYGDVRVSGVMIYRASEGKPMQYVSLLM\n",
      "[['CHAIN', [0, 1164], {'note': 'Uncharacterized protein 041R'}], ['REGION', [421, 441], {'note': 'Disordered'}]]\n",
      "-----\n",
      "042L_FRG3G\n",
      "MFAPPSSLFVPATAPAPSTSGFTIPANLRRDAYVCPFATAEKERKEREQQQPASKGLNHDLAAQEPLHPSLVSRFPSNYRGSFLR\n",
      "[['CHAIN', [0, 84], {'note': 'Uncharacterized protein 042L'}], ['REGION', [0, 21], {'note': 'Disordered'}], ['REGION', [40, 84], {'note': 'Disordered'}], ['COMPBIAS', [74, 84], {'note': 'Polar residues'}]]\n",
      "-----\n",
      "042R_IIV6\n",
      "MATLQQAQQQNNQLTQQNNQLTQQNNQLTQRVNELTRFLEDANRKIQIKENVIKSSEAENRKNLAEINRLHSENHRLIQQSTRTICQKCSMRSN\n",
      "[['CHAIN', [0, 93], {'note': 'Uncharacterized protein 042R'}], ['REGION', [0, 21], {'note': 'Disordered'}], ['COILED', [0, 76], {}]]\n",
      "-----\n",
      "043L_IIV6\n",
      "MDLINNKLNIEIQKFCLDLEKKYNINYNNLIDLWFNKESTERLIKCEVNLENKIKFNQKYNSDTIKIMNILFLICSDGVFGKIENNDVKPLTDEDEKICVKFGYKIMIGCLNDIPI\n",
      "[['CHAIN', [0, 115], {'note': 'Uncharacterized protein 043L'}]]\n",
      "-----\n",
      "043R_FRG3G\n",
      "MEEVDGCAGPNSEAGALTAGALTAGAFAVTAGAGVAGAGVAGVGWCSWCSWCSWCWCSWCSWCWCSWCWCSWCWCSWCWCSWCWCSWCWCSWCWCSWCLSKGWEDRGGLEGCKSCKGWCLCSHCWCWCSWCWCSWCSWCLSKGWEDRGGLEGCKSCKGWCLCSHCRCWSIN\n",
      "[['CHAIN', [0, 170], {'note': 'Uncharacterized protein 043R'}], ['TRANSMEM', [24, 44], {'note': 'Helical'}]]\n",
      "-----\n",
      "044L_IIV3\n",
      "MPLSVFAEEFAEKSVKRYIGQGLWLPCNLSDYYYYQEFHDEGGYGSIHRVMDKATGNEVIMKHSYKLDFSPGILPEWWSKFGSLTDDLRERVVSNHQLRVSREAQILVQASTVLPEMKLHDYFDDGESFILIMDYGGRSLENIASSHKKKITNLVRYRAYKGNWFYKNWLKQVVDYMIKIYHKIKILYDIGIYHNDLKPENVLVDGDHITIIDFGVADFVPDENERKTWSCYDFRGTIDYIPPEVGTTGSFDPWHQTVWCFGVMLYFLSFMEYPFHIDNQFLEYALEGEKLDKLPEPFAQLIRECLSVDPDKRPLTSLLDRLTELHHHLQTIDVW\n",
      "[['CHAIN', [0, 334], {'note': 'Putative serine/threonine-protein kinase 040L'}], ['DOMAIN', [32, 328], {'note': 'Protein kinase'}], ['ACT_SITE', [195, 196], {'note': 'Proton acceptor'}], ['BINDING', [38, 46], {'ligand': 'ATP', 'ligand_id': 'ChEBI:CHEBI:30616'}], ['BINDING', [61, 62], {'ligand': 'ATP', 'ligand_id': 'ChEBI:CHEBI:30616'}]]\n",
      "-----\n",
      "044R_FRG3G\n",
      "MVVRLAVRANMPKDSLARDSLPKDSLARDFLSDKTSPTDGTQSSDRYLLKIVTAVDYVHLT\n",
      "[['CHAIN', [0, 60], {'note': 'Uncharacterized protein 044R'}]]\n",
      "-----\n",
      "044R_IIV6\n",
      "MYLYQKIKNCLLLTMYQKKNKSHMYDILQSYLYYQKPIPKNLYSHPKKNLYLNIHHYKNINKDLM\n",
      "[['CHAIN', [0, 64], {'note': 'Uncharacterized protein 044R'}]]\n",
      "-----\n",
      "045L_FRG3G\n",
      "MDDVEYRTEFSARERAGGDIEEGLELFGPATFKGMEGDPVQRFYNGIESAGRNLIRDGHIKLNKQEQTRLLSSVLRITYPNYKNPMGTVLGFYVTDGGRGPIDKGRLSHVQSFMEEVTDMDLRDLIRYCRLWLALK\n",
      "[['CHAIN', [0, 135], {'note': 'Uncharacterized protein 045L'}]]\n",
      "-----\n",
      "045R_IIV3\n",
      "MYKCSQGAMNTEKVMEKFVIQSRFREMYPDKAKAIAGMTVPARYADSVEDMVAFANEKIRVQKAKVEAEKNARQAMGAPAKFDKYGKYKY\n",
      "[['CHAIN', [0, 89], {'note': 'Uncharacterized protein 045R'}]]\n",
      "-----\n",
      "046L_FRG3G\n",
      "MYSVRNSGCSVGCSPRQGASPIMFGPSLGAMLSAPVVRASAPVVRASSPVVKRKSLVKRKSPVKRSPLKKRSQMRTSPCEA\n",
      "[['CHAIN', [0, 80], {'note': 'Uncharacterized protein 046L'}], ['REGION', [45, 80], {'note': 'Disordered'}], ['COMPBIAS', [50, 72], {'note': 'Basic residues'}]]\n",
      "-----\n",
      "047L_FRG3G\n",
      "MHTISDWLNLNAPVRCFHVRQLSESEWRFTVNDTIRVVASVDGPWTVDAKGVEDLKMHKLYVPGPAKCWTRARDKAMAAALAEAVSESETCAADIVRPAVAKNTPRRPVVKRRVDAVKPAAPDNLESWTKDDWYELDL\n",
      "[['CHAIN', [0, 137], {'note': 'Uncharacterized protein 047L'}]]\n",
      "-----\n",
      "048L_FRG3G\n",
      "MTAKTLDPSDYNVRDDSTTGMFTPVDRFVCDPESDRIIVRKIPPEWTIGNSMRFVHFTKEFTQTFDPSESPSNIVRHTNGKKK\n",
      "[['CHAIN', [0, 82], {'note': 'Uncharacterized protein 048L'}]]\n",
      "-----\n",
      "049L_FRG3G\n",
      "MAAPKAEGEDKPKRVRKSRAKPKPETKEVKKPKSKEFCTADDSSDDYNEVKPSPAMIALMAVKEIPESEDVPDKSDSEAEAPVPAIVKKRRTPPKKAESSDDKKLDEATGEQVIDEDALSKLTIQTLKGMCKTRNLKISGNKAALVQRLIEADGIAHIIPTTATVVQKVKKTKRPAVFSKVDSELKLIPCPGREHMLMDEATGLVFLDEDPSTAVGFIEHGEVFGLDSEHMTVCKNMGIRYSWTEDYLC\n",
      "[['CHAIN', [0, 248], {'note': 'Putative SAP domain-containing protein 049L'}], ['DOMAIN', [118, 152], {'note': 'SAP'}], ['REGION', [0, 109], {'note': 'Disordered'}], ['COMPBIAS', [0, 11], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [21, 37], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [94, 106], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "049L_IIV6\n",
      "MDKIEELKIEELKIEIPQRKTKFFHDSENSDKRDEEETLNPTITSKAKILIKSKNFWIETLIFVISVFGALCVAFGIMLIGFLLWLVSNTISILYFIKQKQYPLSLQQMVFLITTCIGVYNNV\n",
      "[['CHAIN', [0, 122], {'note': 'Transmembrane protein 049L'}], ['TRANSMEM', [66, 86], {'note': 'Helical'}], ['TRANSMEM', [103, 120], {'note': 'Helical'}]]\n",
      "-----\n",
      "049R_IIV3\n",
      "MLRIENTVCKSACRVDSATAQPVYSSFDGENFKAEIHSKLDSFERKLNASPTYRDEEGGGNPEHYETLSQEINDLQSQIENLSLEVENLQGSSSSPSNVAAALAELSQSIRTIKEQLEANRKERYNLTVTVANLTAAVNAAKKTGSESTTATATTTTNYETQLKAFEAQIKALDNQLQTQKNLVQTTSVEAKNDRDSLRKTIEVIRLTVKTLQDQVESQTGPKKRRKSPIENQPTAGSELATLTTNLTFLTQRVEKLSQGVATHTTAMFTLEETMKKVHTTLQEATASNTNNIDAIRTRVQELADKIALFDQVQYSVGYEMAKKNPDSTKLRTDLDSAISTVNEEKKSLLTVKDSVQSLKTQLDELKRTLENDGDVSSLRQTVHDMASSIRDETATIYNKINALEEGLKRGGQTTTTPLTQLQTRVEEIDKTIVKWNNQHGEWTTRLNKLEAGVSNNQTLMNRFIQQVNGDVNPLKELPAELETFKMTITNTWAQLNKKFLDFSAKTDTSVDNFTKKFTEIHPQIASLVDKMDQQIRDNPHTTEKLMDEIRQLKSAMTRLGTQSSGKPIFSINTKSSYNEKSKKTIFGHPGIIFPETVKISSIYITLAAKEADGKEDARLFELTATSTHNNITSTIKQFEKKCTEETILEDYNPPLVIDAQTKLVLSCNQKVFGVAIFTLQYS\n",
      "[['CHAIN', [0, 682], {'note': 'Uncharacterized protein 049R'}], ['REGION', [212, 236], {'note': 'Disordered'}], ['COILED', [61, 123], {}], ['COILED', [154, 258], {}], ['COILED', [345, 375], {}]]\n",
      "-----\n",
      "050L_FRG3G\n",
      "MQVYSPSKISQQLETFLNSVANGLGHTMSHAMSQTFSETIVASVAKKAPKTSVLAAAQAAMQAEDKVSKPKKVKKTKSYADAAPKRVKKVKAPKEDTVVSEPEEAVVEQQEKQQPEKAVVEQQEKQQPEEAVVEQQEKQQPEEAVVEQQEKQQPEEAVVESEQPEQPEQPERQQQAQPERQQQAQPERQQQAQPEEAEDAEQEPVEQPTAKPKKVRKTQTESEDKPKRG\n",
      "[['CHAIN', [0, 228], {'note': 'Uncharacterized protein 050L'}], ['REGION', [60, 228], {'note': 'Disordered'}], ['COMPBIAS', [108, 127], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [165, 193], {'note': 'Low complexity'}], ['COMPBIAS', [194, 203], {'note': 'Acidic residues'}], ['COMPBIAS', [217, 228], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "051R_FRG3G\n",
      "MTVRITTTTGYPHGLLNPTVPLPMSRYTLVQETAQDILYTPMGKELALMGTCTKTECKYVRMQEDAHLLAEASKKALTMRLEQNPELKDLLASSGSQPIVYGDTRLVAHLTLLRGVSVLSPEGMVFSEDAMRKLYKGTIDMFVSDPSSLLNVDRATLTLESLRAMVKASGNWPSTSATAAVAIPEAPVTGVSVSDKVVMEHSASVYATQKMDFERSLLIRHLLAMDPAAEADVSHLVSRMDARTRAASSRLAAMYHDGLLDSAVTDGLVPPDQRLLEPMSTPSTPEVHAPQDGMSFEVPHVLTFAGGPVKVDDHVYDTPLHYAYNLAIRRMFADFGEGDLDDVHVSQVSVIYSDMLDKWIDAMYPQTLWRLMTEKFSGNQSCLAVLLSTDGADVKWTGRTEEESFLISDMMGQIKTGWIRSGPPSSSPLSSADIAGTDFFYGWLSYMSRTYATALNVISEMTLARLLDLPDIPEEVRQPTDREQAALGSDYVRSAWRVCYSEFVHKFEGKNLFASVDYCVKTHLKALKVSRDSVTGTAKTLSAKGYGTLIALPVIRLAMKS\n",
      "[['CHAIN', [0, 560], {'note': 'Uncharacterized protein 051R'}]]\n",
      "-----\n",
      "052L_FRG3G\n",
      "MVKYVVTGGCGFLGSHIVKCILKYAPEVTEVVAYDINISHIMTMWSSKLKVVRGDVMDVMALAKAVDGADVVIHTAGIVDVWYRHTDDEIYRVNVSGTKNVLMCCINAGVQVLVNTSSMEVVGPNTTSGVFVRGGERTPYNTVHDHVYPLSKDRAEKLVKHYTGVAAAPGMPALKTCSLRPTGIYGEGCDLLEKFFHDTVNAGNVAYGGSPPDSEHGRVYVGNVAWMHLLAARALLAGGESAHKVNGEAFFCYDDSPYMSYDAFNAELFEDRGFGYVYVPYWVMKPMAAYNDLKRKFLGCFGVKRSPILNSYTLALARTSFTVKTSKARRMFGYMPLYEWSEAKRRTKDWISTLK\n",
      "[['CHAIN', [0, 354], {'note': 'Uncharacterized protein 052L'}]]\n",
      "-----\n",
      "053L_IIV3\n",
      "MEQYLQAFEFVEEMVVLPKYLSWELYHHLAVLLREKYPKTYKNKGYIFNIKVKSILDNRITPTGQIVLVVMFQSDLYVPQVGHVFTERIRVNSVDDRYQWITIEPLTVFLRSNIPYKPNTLVTVQICSIKMDNTLCFGTILD\n",
      "[['CHAIN', [0, 141], {'note': 'Uncharacterized protein 053L'}]]\n",
      "-----\n",
      "053R_FRG3G\n",
      "MGAAESINTVNIVTKAYAKIMTTMVTDQDITADQSQVFSIDHVKGDVVIKGDVFTQMLVINLASLMKAIATQSAQDQLIDNIAQQAQAAVSGLNLAQYAYVSNNIDRLITACVQMSTDMRVSCKSKVTMTQSFSVTDVEGDVRVTGVKFNQFANILSSCAMDASVNNDQARDIVSQIKQRGDAKASGLDPTTLIVIIVLVMVGAPMGAGFMAGRRAIGPLLASVGLIGGGAVALGYVPRPVKIEGFSSDPDFTLAQPAATVKGLTFTAAVAKLKSTDGYGALFWKNYDVKGTTAVKLQETLSYFAPAGYDPASWAGVGDSAPPFRIFPGLYQGKGDPGARPRAAYGYAGPVAGPKKGDAYLDGDTGSYYVLGDSWKMRGTISGHQNGRTDYWGTVDPTTTAALTGSERYIWVDPFTLVKSTVWLFTGSPKKWTQQQTAPLDIPLTNTPSDFNVWVYKDDTAVQAVKWSSVGAGVAGAALTASALLMPDSVASSEMSPAVGTGTPAIGTGSPAVGTGFPAHRG\n",
      "[['INIT_MET', [0, 1], {'note': 'Removed'}], ['CHAIN', [1, 521], {'note': 'Putative myristoylated protein 053R'}], ['TRANSMEM', [192, 212], {'note': 'Helical'}], ['TRANSMEM', [216, 236], {'note': 'Helical'}], ['REGION', [502, 521], {'note': 'Disordered'}], ['LIPID', [1, 2], {'note': 'N-myristoyl glycine; by host'}]]\n",
      "-----\n",
      "054L_FRG3G\n",
      "MPLRLCQGRKDRASDPVRDDGSPPRLFVSQVCRRAPKDPQGFQGHRGGQNVGDCSPIFHQEKKQVMRRFYSLCEWK\n",
      "[['CHAIN', [0, 75], {'note': 'Uncharacterized protein 054L'}], ['REGION', [0, 23], {'note': 'Disordered'}], ['COMPBIAS', [7, 21], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "054L_IIV3\n",
      "MASEATVESVETKVESPIVESPVDQGLLESIKNFMDDLAVVTENENFQDYHTIVRRIDETKVKSYNKLVGGFREFFSLNKTALMEGNFEGLIEPHISYKTESGSFFFNFQTTYLETDEANQEIIKEHLNHIWAQIRSENKCPEQLYIDEIFQKLKNKDQLTMDDQLIRDLFTKFQTANFNVTALIRAGCSKAREFLTNNGSQKSSSTFRLIETIENVNVDNFTQMDFMALISKISAIFSESGESNPLNLCLSSLFGGGNTNQPSLTSMFPFPTPPLPDNVLLDNLDQLTLEQQSETTGDDDHHSFEPEK\n",
      "[['CHAIN', [0, 308], {'note': 'Uncharacterized protein 054L'}]]\n",
      "-----\n",
      "055L_FRG3G\n",
      "MAKLLRLNAIDGDMPGAGEADLFTLAPGGKAYVPFAWGSRVLGCKPPPAHGAARERGSVSLRPHQKGVLKEAWGHVTSKGYCMLKCPPGFGKTFMALELWRRLGLPALVLTNRRVLATQWRDSATRFLPDSRVFTSGTPPPDALPRDLYVTGPASLRNRRIKAKDSPAKFLLIVDEAHQLTSPVSCRVLLSVRPSHLLGLSATPMRYDDYHAALGAFFGREDSTVDRVDPRPHEVEILSTGVHIEPEFSKITGKMDWNSVIKAQSDNPERDAALADRMLLRPDVKWLVLCKRVDHVKRMAETLSSRSGKKVDVLHGSKDEWDRDAWCVVGTYSKAGTGFDACERTGLCLAADVDRYFEQCLGRLRANGGTVLDPVDDLGVLRKHSKNREAVYIAAGCTIKKTKCDASRPSQSTPTPTGSSQPAPRTRRPQR\n",
      "[['CHAIN', [0, 430], {'note': 'Putative helicase 055L'}], ['DOMAIN', [72, 221], {'note': 'Helicase ATP-binding'}], ['REGION', [402, 430], {'note': 'Disordered'}], ['MOTIF', [174, 177], {'note': 'DEAH box'}], ['COMPBIAS', [406, 423], {'note': 'Low complexity'}], ['BINDING', [85, 92], {'ligand': 'ATP', 'ligand_id': 'ChEBI:CHEBI:30616'}]]\n",
      "-----\n",
      "055R_FRG3G\n",
      "MLPQNSQVVHGVQDGPPVGPQPAQALLKVPVDVRRQAQAGPLAGVEPRPRLGVGAHHTPGVPVPLILGAVQHVHLLPGPRGQCLGHPLDVVHPLAQHQPLYVGPEEHPVGQGGVPLGVIGLGLDHRVPVHLARDLAKLGLYVHARAEYLHLVGPGVDPVHRAVLPAEKGPQCSVVVVVPHGCSAQTQQVRGPHRQEDPTRHGGRQLVGLIHNQEKFCGRVLGLDPPVSQRSRTRHIQVPGQGVGRRGAGCKDPRVRKEPGRRVPPLSRQHPPVCQDEGGQPQPPPQLQGHEGLAEAGRALEHAVPLGGDVAPGLLQDPFLVRSERDGAPLPGCAVSGRRFASQDPGTPGEGDVGLSPGREGKQVRFPRAGHVSIYRVEP\n",
      "[['CHAIN', [0, 378], {'note': 'Uncharacterized protein 055R'}], ['REGION', [0, 19], {'note': 'Disordered'}], ['REGION', [226, 289], {'note': 'Disordered'}], ['REGION', [330, 370], {'note': 'Disordered'}], ['COMPBIAS', [6, 19], {'note': 'Low complexity'}], ['COMPBIAS', [248, 260], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "056R_FRG3G\n",
      "MGVYSPAPRTPRGPWNIRIRFLSWSNSFLLEVKKNYGDVYLCDVCPVRPPGLQAPREQPVLHDRKVLHLYGQDSGVRDVQEVLWNPVSHQEVRRDNHGVLHGRRRARVRQAEEGRRRRQEGHRFRDWERLHQRVEGCPGLQGHGF\n",
      "[['CHAIN', [0, 144], {'note': 'Uncharacterized protein 056R'}]]\n",
      "-----\n",
      "056R_IIV6\n",
      "MEQKIDKKNSYSFGITSSTTVHVLGEVVAIGGILYYTHSQVNQLNTKIASLEKQILDLTNILKHLSPHSFQQLQSSPTTQSTPPLPQSTPQSQQSQQSAVLPQRPSFLGGSTPKESQSFPGVETRSLGEKTTRGKLKSPHPSQIPVTQENFHYPYQTMNKTMRWEFLKPVESDESEDETNCQNGVCTLQKQEKNVTFNGSVEQLKYGNFSPQRSTKTMIGSPSIRPLIPHESIESVSESIESSQDQSFSSRETISGNFKSKDDHQLSESEINQLVSKAIRTKK\n",
      "[['CHAIN', [0, 282], {'note': 'Uncharacterized protein 056R'}], ['TRANSMEM', [12, 34], {'note': 'Helical'}], ['REGION', [69, 145], {'note': 'Disordered'}], ['REGION', [233, 268], {'note': 'Disordered'}], ['COILED', [36, 65], {}], ['COMPBIAS', [70, 97], {'note': 'Low complexity'}], ['COMPBIAS', [233, 251], {'note': 'Low complexity'}]]\n",
      "-----\n",
      "057L_IIV3\n",
      "MFKIYRTSCMGQHQSQFLHSGTVVQTVDGVTTTSFFQPCLVFPFSIEIISISLVSLNTTNETKLIKMSIMENSELVDYNESAYTLAHLPGKQMTYLKYPAPFTIRQHQPFFFVHHGDLGDASLTLEYRIK\n",
      "[['CHAIN', [0, 129], {'note': 'Uncharacterized protein 057L'}]]\n",
      "-----\n",
      "057R_FRG3G\n",
      "MAMVSNVKYFADALQGTQGKVGTFTVLGENVFFKRGDGTDTVCGLEMVAGRILRARSDVHFCEPKYFVEMDDGEKVCSFELLDCKPLGSMAPGRKGKKSVGSVTQYLSGLYQTFAAAAAAHSVGVVHSDLHTGNVMLCPEPVSHYVYNLGGGEMLSLETNGVRAVVVDLGMARIPGKNTVACDIFVHVGHVVNGRPDYAADVRTLTLGSCYDMVMMCASGKPSLEERMLCYEVMAAYNNLFAGVCAPSKGGWFVDHYPSMCAVMEATIPDSVASRGGGSWLLAVANMCKLLVPRPYVKRACGKEKAHAMWMTLFTELGLTAKKSISKVDMVDAVQRLRAIADGSEIPPASLMKAACAVGLLTASVAEACYEKVEEIKASHVGMLRWKDALDAWVRLPVRCSGSVPKLGSTVILHTESGTEETVVTQSMLRQIVKTREALDMAQAASDAVWTDTAYYEADDELMKGAHEESAEDFATSFLKGGTTGPIAKRCRLILKSL\n",
      "[['CHAIN', [0, 497], {'note': 'Putative phosphotransferase 057R'}]]\n",
      "-----\n",
      "058R_FRG3G\n",
      "MEIRDTTVGLDRPVQSGAWDPGATREQLALAGISGRCDLGGRDEDLWSRKSQKDETFKDCERRRGGEAPRLLCDKWRRDREAEAVRRRVQSQRRQGGREDPDGPARVRGRRRRVAVCGHSPDRGPAREEHKGPAGGSDHKDLGHHKGPEVGSGLRGPQARQVCGRSRGHEVQRVQKGLRQGVQGHGCREVALDARLCLVRVLGRSGQKGRHGRLSVASRRLDGQEKVEKRGLFSLGA\n",
      "[['CHAIN', [0, 236], {'note': 'Uncharacterized protein 058R'}], ['REGION', [0, 22], {'note': 'Disordered'}], ['REGION', [81, 166], {'note': 'Disordered'}], ['COMPBIAS', [95, 104], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [119, 147], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "059L_FRG3G\n",
      "MPAQKRMSRYAYLVNKVAGPTLCGVFYGKYVEASDQAVSTCMAWFKIVVITKRVSAREWQEKGATVVTFRDLTRGTAPKAVECDLLYVRDFNPNTSQCWLPMLEGYKRTRTWVQIGHCNFESYGPHVMKKQLFGDFSYVPFVGYGSDVLALPETDEILEYDTSPFERFVHAVQRLNPKGESVWTHLGPDVAFAEMSIDAFNAVRESSPSPMSACMMDKTAVPQSLSEMVADYDWDHNEALRETPYVSSACAAMSRLTSSDPSLVVAVHEPKLQKYLEQFTGAKVVTYGKMLDSRMVPKRVVIFQAHPINGCKSGRKLMHMWRSMPGKPKITVIKPRQEWSAYYCNPLKRVGL\n",
      "[['CHAIN', [0, 351], {'note': 'Uncharacterized protein 059L'}]]\n",
      "-----\n",
      "060L_IIV6\n",
      "MSKNKSPLLNESEKMMSEMLPMKVSQSKLNYEEKVYIPTTIRNRKQHCFRRFFPYIALFQIIMLIILLILYFCFPNLFYSTNFNTNFNTSLLQNNSIETKLNSIPPQNNSQKTEVPIILNYTTQKTEVTEPIIINNTTEEIETQTIMIPKSTDQTQTIISAKTTAIISPPETSETIAQVLKNSDKREHDDEELSFTTEMETITTETETSSTIPHLRSLPIKSESSMETTSEETDEE\n",
      "[['CHAIN', [0, 235], {'note': 'Uncharacterized protein 060L'}], ['TRANSMEM', [51, 71], {'note': 'Helical'}], ['REGION', [182, 235], {'note': 'Disordered'}], ['COMPBIAS', [195, 211], {'note': 'Low complexity'}]]\n",
      "-----\n",
      "061L_FRG3G\n",
      "MWRWSNIYRIGTLSRYYRYAEHSKDTTMSIFSGGVLPNTPSLEKSLYYDAFPYGLTCKNS\n",
      "[['CHAIN', [0, 59], {'note': 'Uncharacterized protein 061L'}]]\n",
      "-----\n",
      "061R_IIV6\n",
      "MNDRGIIICVGIAFLIFIFLWAYFKNKKVKTGNYGNYSYYDKLTDRLDAKLGHDSSFMIQHYLDPENCPLYGDLPTLSQLVNLKQNNPSKYEELVNDGTVKRAVNIKALKTWLLLGSSNVDKFKHFMNQVNQLNIDKVDMLNLENALKYQLSKTQ\n",
      "[['CHAIN', [0, 154], {'note': 'Uncharacterized protein 061R'}], ['TRANSMEM', [4, 24], {'note': 'Helical'}]]\n",
      "-----\n",
      "062L_IIV3\n",
      "MNPEGENELQVYLRDGDLRVMALGLRHGLRRGLRRGLRRGLRRTGDLRDGDLRLDGDLRDGDLDLLDGDLRDGDLRLDGDLDLLEGDLRDGDLDLLDGDLHRDGDVDRLEGDLRDGDLRLDGDFLVLPHTTAWVFLPNSFNFLEVQTLFAPFSYWLHINGIVFINSKILFIKSPQLPSYFAMALRQCKSSWLGHTSPANRFCIVVQINDQLGLKNSINSRRSALWSGGTASC\n",
      "[['CHAIN', [0, 231], {'note': 'Uncharacterized protein 062L'}]]\n",
      "-----\n",
      "062L_IIV6\n",
      "MKKNMILFFGILKKLLICILKMEIKCWLTSDIVYFDSELALTVKQFFMKKNNTILDKIAAHCYGLIMQKISQPVTFKNYIYIWRAVLFADCTIKTNKTPDTQNIINLSQNATEEVKIIIDELIDCFKNKNNFKEEEYKPNLDLLNSYIKNIKKFITENKTSKFIFDKDWEILINNIWMN\n",
      "[['SIGNAL', [0, 25], {}], ['CHAIN', [26, 178], {'note': 'Uncharacterized protein 062L'}]]\n",
      "-----\n",
      "064L_IIV3\n",
      "MDAVKHPVSSLAPHRLGKGFYFMVNSKQIIKKFYNKSFLECSATQETSTTPHDRDNMDQAKLCSLVFEIFKKQTRLLAHLVLEEAIDLNNDLLFAVIYFNDEAILNSLVRHLYKYKPYYCDFTVRAQELDLVVHLDLGHCIDRLKSFIDPDTACFVLCSNLSQLTGLNCLKRVLKHKIIQKSYHLYLLLKTSHKVQQQWPDPVHVVEKYVTKRMVSYALTDNNPLLLAIVLDRLLVKLPTDDFAPLIVGIIESNRFKVECLPTLLQYHNRVKTTTKPIRI\n",
      "[['CHAIN', [0, 279], {'note': 'Uncharacterized protein 064L'}]]\n",
      "-----\n",
      "064R_FRG3G\n",
      "MQNFGAQEMRKGRLAFVRLSKLETLQNLIDKMLAERVFNKGEAADILESNDIRADIARALIDSVTKKGDVACSLFAGAIARQDVVLADAMGISQV\n",
      "[['CHAIN', [0, 94], {'note': 'Caspase recruitment domain-containing protein 064R'}], ['DOMAIN', [0, 93], {'note': 'CARD'}]]\n",
      "-----\n",
      "065L_FRG3G\n",
      "MRGLEAPGAVGPTGPSGAPGSQGPDGDVGGMGPEGPKGDDGPVGPKGPQGAAIF\n",
      "[['CHAIN', [0, 53], {}], ['DOMAIN', [6, 50], {'note': 'Collagen-like'}], ['REGION', [0, 53], {'note': 'Disordered'}], ['COMPBIAS', [25, 34], {'note': 'Gly residues'}], ['COMPBIAS', [41, 53], {'note': 'Low complexity'}]]\n",
      "-----\n",
      "065R_IIV6\n",
      "MNIPKTCFQIHNKIQVQNYLIRINLNIFLIYHFSPIYCPYLFLFTVFFNSLINLI\n",
      "[['CHAIN', [0, 54], {'note': 'Uncharacterized protein 065R'}], ['TRANSMEM', [26, 46], {'note': 'Helical'}]]\n",
      "-----\n",
      "066L_FRG3G\n",
      "MPFYICSDPDPKRTVRGPRFTVPDPKPPPDPAHPLDDTDNVMTAFPKFKPYGFSVYNPWDPIFLSMFGRAGRNGAKGPRGRRRSPRPPGGSSMTPGDDGNQGPRGPGEQRDQPDQMDPLVHPVTSVRSGPWERLGLRGRGESGVLRETLEMWAGQEGLISRVRDDPRESEVRPVTGVQTVWPE\n",
      "[['CHAIN', [0, 182], {'note': 'Uncharacterized protein 066L'}], ['REGION', [0, 43], {'note': 'Disordered'}], ['REGION', [67, 136], {'note': 'Disordered'}], ['REGION', [162, 182], {'note': 'Disordered'}]]\n",
      "-----\n",
      "068R_FRG3G\n",
      "MWIHFPVRNYIHLSHTILHYPCPDNAGHVCGHRNVSRNGLKDGLGVGVVSGGVDRTVYDRDTYPHPIKESDEDLGGGGVGSRERVGPVGGDHCHG\n",
      "[['CHAIN', [0, 94], {'note': 'Putative uncharacterized protein 068R'}], ['REGION', [56, 94], {'note': 'Disordered'}], ['COMPBIAS', [56, 68], {'note': 'Basic and acidic residues'}], ['COMPBIAS', [80, 94], {'note': 'Basic and acidic residues'}]]\n",
      "-----\n",
      "069L_IIV6\n",
      "MSDKIDNQIVKVENTNNGGLRAIFNLDGVTLDTPIMGTWDKPVFFGKEIAEFLGFKKPKDALQKHVKPKYKTTLSKVLEKKLDTEPVSYNEGKRVLLYKEGVVELIKKTRLVGIENKIDALIEAFELNLNVVHQTKEQEHIGAILKVFTHERTVTQYKVKNYKIDLYFLDYKVAVECDEFGHRDRNGESEREYNIIRELGCTFYRFNPDSKNFCIYQIIGDLVRLLYKDH\n",
      "[['CHAIN', [0, 229], {'note': 'Putative Bro-N domain-containing protein 069L'}], ['DOMAIN', [19, 138], {'note': 'Bro-N'}]]\n",
      "-----\n",
      "069R_FRG3G\n",
      "MERSATLEMLNVHKPDARQTGDILSRYANALWPPALAYAASVAAGYVFTAGPHGCRCGKAMTEAARVGVFLGVLCALYNWMGSGDSFA\n",
      "[['CHAIN', [0, 87], {'note': 'Transmembrane protein 069R'}], ['TRANSMEM', [29, 49], {'note': 'Helical'}], ['TRANSMEM', [66, 86], {'note': 'Helical'}]]\n",
      "-----\n",
      "070R_FRG3G\n",
      "\n",
      "[]\n",
      "-----\n"
     ]
    }
   ],
   "source": [
    "# Sanity check!\n",
    "# Check that our uniprot data entries behave as expected\n",
    "for e_i, entry in enumerate(uniprot_data):\n",
    "    print(entry.entry_ID)\n",
    "    print(entry.sequence)\n",
    "    print(entry.features)\n",
    "    print('-----')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "fIPl7BmeRQi_"
   },
   "source": [
    "### Filter and Format\n",
    "Identify the samples we are interested in and convert sequence data and labels to one-hot values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "LOm0Ax5FwSvd"
   },
   "outputs": [],
   "source": [
    "feature_labels = [\n",
    " 'DNA_BIND',\n",
    " 'REGION-Disordered',\n",
    " 'TOPO_DOM-Cytoplasmic',\n",
    " 'TOPO_DOM-Extracellular',\n",
    " 'TOPO_DOM-Lumenal',\n",
    " 'TRANSIT-Mitochondrion',\n",
    " 'TRANSMEM-Helical',\n",
    " 'ZN_FING'\n",
    "]\n",
    "\n",
    "seq_vocab = ['-','A','B','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','U','V','W','X','Y','Z']\n",
    "\n",
    "def get_label_index(feature):\n",
    "    for k,v in feature[-1].items():\n",
    "        feature_string = f\"{feature[0]}-{v}\"\n",
    "        for idx, label in enumerate(feature_labels):\n",
    "            if label in feature_string:\n",
    "                return idx\n",
    "    return None\n",
    "\n",
    "def get_samples(entry, feature_labels, max_l=20):\n",
    "    samples = []\n",
    "    for feature in entry.features:\n",
    "        label_idx = get_label_index(feature)\n",
    "\n",
    "        if label_idx is not None:\n",
    "            seq_slice = entry.sequence[slice(*feature[1])][:max_l]\n",
    "\n",
    "            if len(seq_slice) < max_l:\n",
    "                seq_slice += '-'*(max_l-len(seq_slice))\n",
    "\n",
    "            samples.append((seq_slice, label_idx))\n",
    "\n",
    "    return samples\n",
    "\n",
    "def generate_onehot(size, idx):\n",
    "    onehot = [0 for _ in range(size)]\n",
    "    onehot[idx] = 1\n",
    "    return onehot\n",
    "\n",
    "def seq_to_onehot(seq, vocab):\n",
    "    seq_onehot = []\n",
    "    for c in seq:\n",
    "        symbol_onehot = generate_onehot(len(vocab), vocab.index(c))\n",
    "        seq_onehot += symbol_onehot\n",
    "    return seq_onehot\n",
    "\n",
    "# Store one-hot featured samples in these!\n",
    "sample_seqs = []\n",
    "sample_labels = []\n",
    "\n",
    "for ei, entry in enumerate(uniprot_data):\n",
    "    samples = get_samples(entry, feature_labels)\n",
    "\n",
    "    if samples is not None:\n",
    "        for seq, label in samples:\n",
    "            sample_seqs.append(seq_to_onehot(seq, seq_vocab))\n",
    "            sample_labels.append(generate_onehot(len(feature_labels), label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "0EwtSzvXRPjZ"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "500 25\n",
      "[0, 0, 0, 0, 0, 0, 1, 0]\n"
     ]
    }
   ],
   "source": [
    "for seq, label in zip(sample_seqs, sample_labels):\n",
    "    print(seq)\n",
    "    print(len(seq), len(seq_vocab))\n",
    "    print(label)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "MYqP4rNRFluf"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_438/3113513996.py:1: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  sample_seqs = torch.tensor(sample_seqs)\n",
      "/tmp/ipykernel_438/3113513996.py:2: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  sample_labels = torch.tensor(sample_labels)\n"
     ]
    }
   ],
   "source": [
    "sample_seqs = torch.tensor(sample_seqs)\n",
    "sample_labels = torch.tensor(sample_labels)\n",
    "\n",
    "uniprot_corpus = (sample_seqs, sample_labels, seq_vocab, feature_labels)\n",
    "\n",
    "# Don't mess with this unless you want to overwrite precious data!\n",
    "# torch.save((sample_seqs, sample_labels, seq_vocab, feature_labels), 'uniprot_corpus.pt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "rzVO2-knZqOT"
   },
   "outputs": [],
   "source": [
    "# Let's convert our one-hot data back into text data\n",
    "# (just another check)\n",
    "\n",
    "fun_hot_seqs = sample_seqs.reshape((-1, 20, len(seq_vocab))).argmax(2)\n",
    "fun_hot_labels = sample_labels.argmax(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "id": "-acvqiRaXr-y"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VMFFVAGAVLVAILIST--- TRANSMEM-Helical\n",
      "YVPPPPPPKPTPPTPPTPPT REGION-Disordered\n",
      "GTGYESDSDPENEHFDDESF REGION-Disordered\n",
      "YNSEDEDFEYDSDSEDDDSD REGION-Disordered\n",
      "MLFLGTIGLAVVVGGLMAYG TRANSMEM-Helical\n",
      "GKTPSSGTSFHTASPSFSSR REGION-Disordered\n",
      "MQNPLPEVMSPEHDKRTTT- REGION-Disordered\n",
      "RFTKPSSSVAKSTSPSLRNS REGION-Disordered\n",
      "TVLAFKGEGALALAGLLVMA TRANSMEM-Helical\n",
      "FGVVHSHTPKKKYTSRDSDS REGION-Disordered\n",
      "VKTIAMLAMLVIVAALIYMG TRANSMEM-Helical\n",
      "YKMWFLYALILALIFGVFMW TRANSMEM-Helical\n",
      "APEGMGPHHAASSSHHSAQH REGION-Disordered\n",
      "AVTGSSSNVKIRKSAPARNE REGION-Disordered\n",
      "LVTYGGKDGPSDNEDGPSDD REGION-Disordered\n",
      "MATNYCDEFERNPTRNPRTG REGION-Disordered\n",
      "PRVAAASPCPEFARDPTRNP REGION-Disordered\n",
      "GGASPRRVSPARAFPNRRVS REGION-Disordered\n",
      "GLSPFRSHMRKSPARRSPAR REGION-Disordered\n",
      "SRPSGVSRTSGTSGSSGSSA REGION-Disordered\n",
      "MLQNYAIVLGMAVAVAIWYF TRANSMEM-Helical\n",
      "APPGPNPPKPDPPKPDPPKM REGION-Disordered\n",
      "RQPRVVPVTSSDPEVVDDED REGION-Disordered\n",
      "SFVHKLPTFYTAGVGAIIGG TRANSMEM-Helical\n",
      "HWAGIALYCVGWVTLASVIY TRANSMEM-Helical\n",
      "ILKGSILSCIVISAVWSILE TRANSMEM-Helical\n",
      "LPLISCAMLVSSLAALVALK TRANSMEM-Helical\n",
      "IFTILGAAIIIILAEYVVLP TRANSMEM-Helical\n",
      "IVDGIGLPLLLLGFFILYQV TRANSMEM-Helical\n",
      "QLSFEEDGWTESEPRPVRRE REGION-Disordered\n",
      "GGRGEEFGKPDETPSSASVG REGION-Disordered\n",
      "PLADGATVDSQASENGEKEA REGION-Disordered\n",
      "AIRDHGRVSEARVASPRRSP REGION-Disordered\n",
      "DLNRESDGKPKTKDQLCADI REGION-Disordered\n",
      "ESDESFARGARPKTNKRDSD REGION-Disordered\n",
      "TSNATEASEKTVAKTNGTAA REGION-Disordered\n",
      "LMLGAVIVAAVVVGVRMFM- TRANSMEM-Helical\n",
      "NLLILLSLLLVVVAIMWWLY TRANSMEM-Helical\n",
      "PEDSSCPCPRLPLSPTGPVG REGION-Disordered\n",
      "GDPRHPWDPGAQSGYPRVRE REGION-Disordered\n",
      "WCSIGWSIGALIIFLV---- TRANSMEM-Helical\n",
      "TTPQTTTPEIPSSTEPQELV REGION-Disordered\n",
      "SATSHYPKADDPQRFARSVS REGION-Disordered\n",
      "YTERPYPVSSTQSPTTTQSP REGION-Disordered\n",
      "RLKNGRGNKTMSESDYNTSD REGION-Disordered\n",
      "EAAPPRPPRKSKAPEPTGDK REGION-Disordered\n",
      "MFAPPSSLFVPATAPAPSTS REGION-Disordered\n",
      "EKERKEREQQQPASKGLNHD REGION-Disordered\n",
      "MATLQQAQQQNNQLTQQNNQ REGION-Disordered\n",
      "GAFAVTAGAGVAGAGVAGVG TRANSMEM-Helical\n",
      "ASSPVVKRKSLVKRKSPVKR REGION-Disordered\n",
      "MAAPKAEGEDKPKRVRKSRA REGION-Disordered\n",
      "VFGALCVAFGIMLIGFLLWL TRANSMEM-Helical\n",
      "LSLQQMVFLITTCIGVY--- TRANSMEM-Helical\n",
      "QDQVESQTGPKKRRKSPIEN REGION-Disordered\n",
      "MQAEDKVSKPKKVKKTKSYA REGION-Disordered\n",
      "LIVIIVLVMVGAPMGAGFMA TRANSMEM-Helical\n",
      "IGPLLASVGLIGGGAVALGY TRANSMEM-Helical\n",
      "TPAIGTGSPAVGTGFPAHR- REGION-Disordered\n",
      "MPLRLCQGRKDRASDPVRDD REGION-Disordered\n",
      "KCDASRPSQSTPTPTGSSQP REGION-Disordered\n",
      "MLPQNSQVVHGVQDGPPVG- REGION-Disordered\n",
      "VSQRSRTRHIQVPGQGVGRR REGION-Disordered\n",
      "PGCAVSGRRFASQDPGTPGE REGION-Disordered\n",
      "FGITSSTTVHVLGEVVAIGG TRANSMEM-Helical\n",
      "FQQLQSSPTTQSTPPLPQST REGION-Disordered\n",
      "ESVSESIESSQDQSFSSRET REGION-Disordered\n",
      "MEIRDTTVGLDRPVQSGAWD REGION-Disordered\n",
      "AEAVRRRVQSQRRQGGREDP REGION-Disordered\n",
      "FFPYIALFQIIMLIILLILY TRANSMEM-Helical\n",
      "SDKREHDDEELSFTTEMETI REGION-Disordered\n",
      "GIIICVGIAFLIFIFLWAYF TRANSMEM-Helical\n",
      "MRGLEAPGAVGPTGPSGAPG REGION-Disordered\n",
      "IFLIYHFSPIYCPYLFLFTV TRANSMEM-Helical\n",
      "MPFYICSDPDPKRTVRGPRF REGION-Disordered\n",
      "GRAGRNGAKGPRGRRRSPRP REGION-Disordered\n",
      "RDDPRESEVRPVTGVQTVWP REGION-Disordered\n",
      "VYDRDTYPHPIKESDEDLGG REGION-Disordered\n",
      "ALWPPALAYAASVAAGYVFT TRANSMEM-Helical\n",
      "VGVFLGVLCALYNWMGSGDS TRANSMEM-Helical\n"
     ]
    }
   ],
   "source": [
    "for seq_index, label_index in zip(fun_hot_seqs, fun_hot_labels):\n",
    "    print(''.join([seq_vocab[i] for i in seq_index]), feature_labels[label_index])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct a dataloader\n",
    "Specify a class which partitions the data into train-test splits and retrieves randomized batches. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_corpus = torch.load('uniprot_corpus.pt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "id": "h5JBBtIrMnCD"
   },
   "outputs": [],
   "source": [
    "class UniprotLoader:\n",
    "    def __init__(self, uniprot_corpus, batch_size=64):\n",
    "        self.sequences = uniprot_corpus[0]\n",
    "        self.labels = uniprot_corpus[1]\n",
    "        self.sequence_vocab = uniprot_corpus[2]\n",
    "        self.label_vocab = uniprot_corpus[3]\n",
    "\n",
    "        self.seq_length = self.sequences.shape[1]\n",
    "        self.label_length = self.labels.shape[1]\n",
    "\n",
    "        self.train_index, self.test_index = self.get_train_test_splits()\n",
    "\n",
    "        self.batch_size = batch_size\n",
    "\n",
    "        self.train_iterator = self.new_batch_iterator(self.train_index)\n",
    "        self.test_iterator = self.new_batch_iterator(self.test_index)\n",
    "\n",
    "        self.train_batch_count = len(self.train_index) // self.batch_size\n",
    "        self.test_batch_count = len(self.test_index) // self.batch_size\n",
    "\n",
    "    def new_batch_iterator(self, index):\n",
    "        random_index = index[torch.randperm(len(index))]\n",
    "\n",
    "        for i in range(0, len(random_index), self.batch_size):\n",
    "            yield random_index[i : i + self.batch_size]\n",
    "\n",
    "    def get_train_test_splits(self, r=0.2):\n",
    "        train_index = []\n",
    "        test_index = []\n",
    "\n",
    "        for l_idx in range(len(self.label_vocab)):\n",
    "            class_index = torch.where(self.labels[:, l_idx]==1)[0]\n",
    "            class_size = len(class_index)\n",
    "            class_index_random = class_index[torch.randperm(class_size)]\n",
    "\n",
    "            test_cutoff = int(class_size*r)\n",
    "\n",
    "            test_index.append(class_index_random[:test_cutoff])\n",
    "            train_index.append(class_index_random[test_cutoff:])\n",
    "\n",
    "        return(torch.hstack(train_index).sort()[0], torch.hstack(test_index).sort()[0])\n",
    "\n",
    "    def get_batch(self, dataset='train'):\n",
    "        iterator = self.train_iterator if dataset == 'train' else self.test_iterator\n",
    "        random_index = next(iterator, None)\n",
    "\n",
    "        if random_index is None:\n",
    "            if dataset == 'train':\n",
    "                iterator = self.new_batch_iterator(self.train_index)\n",
    "                self.train_iterator = iterator\n",
    "            else:\n",
    "                iterator = self.new_batch_iterator(self.test_index)\n",
    "                self.test_iterator = iterator\n",
    "\n",
    "            random_index = next(self.train_iterator, None)\n",
    "\n",
    "        return self.sequences[random_index].float(), self.labels[random_index].float()\n",
    "\n",
    "\n",
    "uniprot_loader = UniprotLoader(uniprot_corpus, batch_size=2048)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build your model\n",
    "Construct your sequence classifier and test it in battle\n",
    "\n",
    "**TODO #1:** Implement the SequenceClassifier class. This could look very much like the MNIST classifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "id": "-T9XYaYcWbB3"
   },
   "outputs": [],
   "source": [
    "class SequenceClassifier(nn.Module):\n",
    "    '''\n",
    "    There's nothing here! :O\n",
    "    Consult the MNIST classifier model \n",
    "    or some other resource to construct this class.\n",
    "    '''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO #2:** Instantiate your sequence classifier. You will need to identify the correct hyperparameters, or have hard-coded them directly into the SequenceClassifier class. \n",
    "\n",
    "Be sure to load your model onto the GPU! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "id": "cvLA6W7ar14f"
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sequence_classifier' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[29]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m loss_function = nn.CrossEntropyLoss()\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m optimizer = torch.optim.Adam(\u001b[43msequence_classifier\u001b[49m.parameters(), lr=\u001b[32m0.001\u001b[39m)\n\u001b[32m      5\u001b[39m device = \u001b[33m\"\u001b[39m\u001b[33mHow do you access the GPU?\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      6\u001b[39m sequence_classifier = \u001b[33m\"\u001b[39m\u001b[33mReplace this string!\u001b[39m\u001b[33m\"\u001b[39m\n",
      "\u001b[31mNameError\u001b[39m: name 'sequence_classifier' is not defined"
     ]
    }
   ],
   "source": [
    "loss_function = nn.CrossEntropyLoss()\n",
    "optimizer = torch.optim.Adam(sequence_classifier.parameters(), lr=0.001)\n",
    "\n",
    "device = \"How do you access the GPU?\"\n",
    "sequence_classifier = \"Replace this string!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----Epoch [1/100], Batch [100/298]-----|\n",
      "| Train Loss: 0.0152\n",
      "| Per-label test accuracy: 0.92 | 0.98 | 0.80 | 0.76 | 0.52 | 0.79 | 1.00 | 0.98\n",
      "|----------------------------------------|\n",
      "\n",
      "|-----Epoch [1/100], Batch [200/298]-----|\n",
      "| Train Loss: 0.0145\n",
      "| Per-label test accuracy: 0.92 | 0.97 | 0.80 | 0.76 | 0.51 | 0.83 | 1.00 | 0.98\n",
      "|----------------------------------------|\n",
      "\n",
      "|-----Epoch [2/100], Batch [100/298]-----|\n",
      "| Train Loss: 0.0166\n",
      "| Per-label test accuracy: 0.92 | 0.97 | 0.81 | 0.75 | 0.53 | 0.83 | 1.00 | 0.98\n",
      "|----------------------------------------|\n",
      "\n",
      "|-----Epoch [2/100], Batch [200/298]-----|\n",
      "| Train Loss: 0.0160\n",
      "| Per-label test accuracy: 0.93 | 0.97 | 0.80 | 0.75 | 0.53 | 0.81 | 1.00 | 0.97\n",
      "|----------------------------------------|\n",
      "\n",
      "|-----Epoch [3/100], Batch [100/298]-----|\n",
      "| Train Loss: 0.0156\n",
      "| Per-label test accuracy: 0.92 | 0.97 | 0.80 | 0.76 | 0.53 | 0.79 | 1.00 | 0.98\n",
      "|----------------------------------------|\n",
      "\n",
      "|-----Epoch [3/100], Batch [200/298]-----|\n",
      "| Train Loss: 0.0150\n",
      "| Per-label test accuracy: 0.92 | 0.98 | 0.81 | 0.73 | 0.52 | 0.83 | 1.00 | 0.98\n",
      "|----------------------------------------|\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def train(dataloader, model, loss_function, optimizer, test_freq=100):\n",
    "    device = next(model.parameters()).device\n",
    "    loss_vals = []\n",
    "\n",
    "    for batch_index in range(dataloader.train_batch_count):\n",
    "        model.train()\n",
    "        batch_samples, batch_labels = dataloader.get_batch()\n",
    "\n",
    "        optimizer.zero_grad()\n",
    "\n",
    "        outputs = model(batch_samples.to(device))\n",
    "        loss = loss_function(outputs, torch.argmax(batch_labels.to(device), dim=1))\n",
    "\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "\n",
    "        loss_vals.append(loss.item())\n",
    "\n",
    "        if (batch_index + 1) % test_freq == 0: \n",
    "            test_accuracy = test(dataloader, model)\n",
    "            accuracy_report = ' | '.join([f'{x.item():0.2f}' for x in test_accuracy]) \n",
    "            avg_train_loss = sum(loss_vals) / len(loss_vals)\n",
    "            \n",
    "            report_string = [f'|-----Epoch [%d/%d], Batch [{batch_index+1}/{uniprot_loader.train_batch_count}]-----|\\n',\n",
    "                             f'| Train Loss: {avg_train_loss:.4f}\\n',\n",
    "                             f'| Per-label test accuracy: {accuracy_report}\\n']\n",
    "\n",
    "            report_string.append('|' + ''.join(['-' for _ in range(len(report_string[0])-3)]) + '|\\n')\n",
    "            \n",
    "            yield ''.join(report_string)\n",
    "            \n",
    "\n",
    "\n",
    "def test(dataloader, model):\n",
    "    with torch.no_grad():\n",
    "        model.eval()\n",
    "        totals = torch.zeros(len(dataloader.label_vocab))\n",
    "        correct = torch.zeros(len(dataloader.label_vocab))\n",
    "    \n",
    "        for batch_index in range(dataloader.test_batch_count):\n",
    "            batch_samples, batch_labels = dataloader.get_batch(dataset='test')\n",
    "    \n",
    "            outputs = model(batch_samples.to(device))\n",
    "            predictions = outputs.argmax(dim=1)\n",
    "            true_labels = batch_labels.argmax(dim=1).to(device)\n",
    "            hits = (predictions == true_labels)\n",
    "            correct_label_index, correct_counts = true_labels[hits].cpu().unique(return_counts=True)\n",
    "    \n",
    "            totals += torch.sum(batch_labels, dim=0)\n",
    "            correct[correct_label_index] += correct_counts\n",
    "    \n",
    "        return (correct / totals)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----Epoch [1/100], Batch [100/298]-----|\n",
      "Train Loss: 0.0175\n",
      "Per-label test accuracy: 0.925 0.969 0.809 0.720 0.519 0.862 0.996 0.982\n",
      "--------------\n",
      "\n",
      "|-----Epoch [1/100], Batch [200/298]-----|\n",
      "Train Loss: 0.0174\n",
      "Per-label test accuracy: 0.922 0.971 0.810 0.753 0.513 0.797 0.996 0.978\n",
      "--------------\n",
      "\n",
      "|-----Epoch [2/100], Batch [100/298]-----|\n",
      "Train Loss: 0.0153\n",
      "Per-label test accuracy: 0.910 0.972 0.812 0.738 0.531 0.815 0.996 0.979\n",
      "--------------\n",
      "\n",
      "|-----Epoch [2/100], Batch [200/298]-----|\n",
      "Train Loss: 0.0160\n",
      "Per-label test accuracy: 0.937 0.973 0.811 0.733 0.519 0.770 0.997 0.974\n",
      "--------------\n",
      "\n",
      "|-----Epoch [3/100], Batch [100/298]-----|\n",
      "Train Loss: 0.0163\n",
      "Per-label test accuracy: 0.912 0.977 0.797 0.752 0.504 0.848 0.997 0.975\n",
      "--------------\n",
      "\n",
      "|-----Epoch [3/100], Batch [200/298]-----|\n",
      "Train Loss: 0.0162\n",
      "Per-label test accuracy: 0.922 0.967 0.818 0.740 0.508 0.798 0.997 0.977\n",
      "--------------\n",
      "\n",
      "|-----Epoch [4/100], Batch [100/298]-----|\n",
      "Train Loss: 0.0153\n",
      "Per-label test accuracy: 0.921 0.965 0.811 0.736 0.548 0.867 0.997 0.977\n",
      "--------------\n",
      "\n",
      "|-----Epoch [4/100], Batch [200/298]-----|\n",
      "Train Loss: 0.0156\n",
      "Per-label test accuracy: 0.924 0.971 0.815 0.735 0.522 0.861 0.997 0.979\n",
      "--------------\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[80]\u001b[39m\u001b[32m, line 4\u001b[39m\n\u001b[32m      1\u001b[39m epochs = \u001b[32m100\u001b[39m\n\u001b[32m      3\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m epoch \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[32m100\u001b[39m):\n\u001b[32m----> \u001b[39m\u001b[32m4\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43moutput\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mtrain\u001b[49m\u001b[43m(\u001b[49m\u001b[43muniprot_loader\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msequence_classifier\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mloss_function\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moptimizer\u001b[49m\u001b[43m)\u001b[49m\u001b[43m:\u001b[49m\n\u001b[32m      5\u001b[39m \u001b[43m        \u001b[49m\u001b[38;5;28;43mprint\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43moutput\u001b[49m\u001b[43m \u001b[49m\u001b[43m%\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43mepoch\u001b[49m\u001b[43m+\u001b[49m\u001b[32;43m1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mepochs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[79]\u001b[39m\u001b[32m, line 7\u001b[39m, in \u001b[36mtrain\u001b[39m\u001b[34m(dataloader, model, loss_function, optimizer, test_freq)\u001b[39m\n\u001b[32m      5\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m batch_index \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(dataloader.train_batch_count):\n\u001b[32m      6\u001b[39m     model.train()\n\u001b[32m----> \u001b[39m\u001b[32m7\u001b[39m     batch_samples, batch_labels = \u001b[43mdataloader\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_batch\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m      9\u001b[39m     optimizer.zero_grad()\n\u001b[32m     11\u001b[39m     outputs = model(batch_samples.to(device))\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[31]\u001b[39m\u001b[32m, line 57\u001b[39m, in \u001b[36mUniprotLoader.get_batch\u001b[39m\u001b[34m(self, dataset)\u001b[39m\n\u001b[32m     53\u001b[39m         \u001b[38;5;28mself\u001b[39m.test_iterator = iterator\n\u001b[32m     55\u001b[39m     random_index = \u001b[38;5;28mnext\u001b[39m(\u001b[38;5;28mself\u001b[39m.train_iterator, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[32m---> \u001b[39m\u001b[32m57\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43msequences\u001b[49m\u001b[43m[\u001b[49m\u001b[43mrandom_index\u001b[49m\u001b[43m]\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfloat\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m, \u001b[38;5;28mself\u001b[39m.labels[random_index].float()\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "epochs = 100\n",
    "\n",
    "for epoch in range(100):\n",
    "    for output in train(uniprot_loader, sequence_classifier, loss_function, optimizer):\n",
    "        print(output % (epoch+1, epochs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reframing the problem\n",
    "In this exercise, we were tasked with classifying a fixed length sequence with a well established label. While this makes for an instructive toy problem, it takes little imagination to appreciate how limited in utility our model is.\n",
    "\n",
    "Consider a revised version of this problem: Suppose that instead of receiving a fixed length of sequence to be classified, we are given a full protein sequence. In this case, we are tasked not only with classifying our regions of interest, but identifying the actual bounds of those regions as well. \n",
    "\n",
    "This would make for a much more useful tool than what we have constructed here. However, updating our model to tackle this task is not a trivial matter. \n",
    "\n",
    "Your final **TODO** is to spare some thought about how to approach this problem. Feel free to even take a stab at it here, though I certainly won't hold you to it!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Instructor's Manual\n",
    "All TODO solutions available here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO #1**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SequenceClassifier(nn.Module):\n",
    "    def __init__(self, input_size, hidden_size, output_size, hidden_depth=4):\n",
    "        super(SequenceClassifier, self).__init__()\n",
    "\n",
    "        self.hidden_size = hidden_size\n",
    "\n",
    "        self.input_layer = nn.Linear(input_size, hidden_size)\n",
    "        self.hidden_layers = nn.Sequential(*[nn.Linear(hidden_size, hidden_size), \n",
    "                                            nn.ReLU()] * hidden_depth)\n",
    "        self.output_layer = nn.Linear(hidden_size, output_size)\n",
    "\n",
    "    def forward(self, x):\n",
    "        x = self.input_layer(x)\n",
    "        x = self.hidden_layers(x)\n",
    "        x = self.output_layer(x)\n",
    "        return x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO #2**\n",
    "\n",
    "Note: You may have chosen to hard code the hyperparameters into your model, in which case you will not need to include arguments at all when instantiating your SequenceClassifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "sequence_classifier = SequenceClassifier(uniprot_loader.seq_length, 1024, len(uniprot_loader.label_vocab)).to(device)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO #3**\n",
    "\n",
    "Just have fun!"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [
    "N0zivH3dLry7",
    "lEF2uJmwNzJl"
   ],
   "gpuType": "L4",
   "private_outputs": true,
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
